{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MGYG-HGUT-02492\n",
    "MGYG-HGUT-02506\n",
    "MGYG-HGUT-01302\n",
    "MGYG-HGUT-02438\n",
    "MGYG-HGUT-02478\n",
    "MGYG-HGUT-01346\n",
    "MGYG-HGUT-00258\n",
    "MGYG-HGUT-03694\n",
    "MGYG-HGUT-00060\n",
    "MGYG-HGUT-02321"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading and dealing with the metadata table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_data = pd.read_csv(\"41587_2020_603_MOESM3_ESM.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species representative</th>\n",
       "      <th>MGnify accession</th>\n",
       "      <th>ENA accession</th>\n",
       "      <th>Status</th>\n",
       "      <th>Reference genes</th>\n",
       "      <th>Core genes1</th>\n",
       "      <th>% Core genes1</th>\n",
       "      <th>Accessory genes1</th>\n",
       "      <th>Pan-genome size1</th>\n",
       "      <th>Number of SNVs</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 16</th>\n",
       "      <th>Unnamed: 17</th>\n",
       "      <th>Number of 97% ANI clusters</th>\n",
       "      <th>Number of 99% ANI clusters</th>\n",
       "      <th>Number of genomes (total)</th>\n",
       "      <th>Number of genomes (non-redundant)</th>\n",
       "      <th>Number of genomes (near-complete)</th>\n",
       "      <th>Number of samples</th>\n",
       "      <th>Taxonomy lineage (GTDB)</th>\n",
       "      <th>Taxonomy lineage (NCBI)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GUT_GENOME000001</td>\n",
       "      <td>MGYG-HGUT-00001</td>\n",
       "      <td>ERS3638031</td>\n",
       "      <td>Cultured (human gut)</td>\n",
       "      <td>3182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16351.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia;o__P...</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GUT_GENOME000004</td>\n",
       "      <td>MGYG-HGUT-00002</td>\n",
       "      <td>ERS3638032</td>\n",
       "      <td>Cultured (human gut)</td>\n",
       "      <td>3920</td>\n",
       "      <td>1821.0</td>\n",
       "      <td>46.45</td>\n",
       "      <td>34621.0</td>\n",
       "      <td>36442.0</td>\n",
       "      <td>428560.0</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>162</td>\n",
       "      <td>134</td>\n",
       "      <td>327</td>\n",
       "      <td>358</td>\n",
       "      <td>308</td>\n",
       "      <td>120</td>\n",
       "      <td>308</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia;o__L...</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GUT_GENOME000008</td>\n",
       "      <td>MGYG-HGUT-00003</td>\n",
       "      <td>ERS3638033</td>\n",
       "      <td>Cultured (human gut)</td>\n",
       "      <td>2690</td>\n",
       "      <td>1522.0</td>\n",
       "      <td>56.58</td>\n",
       "      <td>19714.0</td>\n",
       "      <td>21236.0</td>\n",
       "      <td>516967.0</td>\n",
       "      <td>...</td>\n",
       "      <td>244</td>\n",
       "      <td>676</td>\n",
       "      <td>42</td>\n",
       "      <td>561</td>\n",
       "      <td>1178</td>\n",
       "      <td>1000</td>\n",
       "      <td>593</td>\n",
       "      <td>1003</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>d__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GUT_GENOME000010</td>\n",
       "      <td>MGYG-HGUT-00004</td>\n",
       "      <td>ERS3638034</td>\n",
       "      <td>Cultured (human gut)</td>\n",
       "      <td>3634</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>58094.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>24</td>\n",
       "      <td>22</td>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia;o__O...</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GUT_GENOME000017</td>\n",
       "      <td>MGYG-HGUT-00005</td>\n",
       "      <td>ERS3638035</td>\n",
       "      <td>Cultured (human gut)</td>\n",
       "      <td>3755</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia;o__P...</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species representative MGnify accession ENA accession                Status  \\\n",
       "0       GUT_GENOME000001  MGYG-HGUT-00001    ERS3638031  Cultured (human gut)   \n",
       "1       GUT_GENOME000004  MGYG-HGUT-00002    ERS3638032  Cultured (human gut)   \n",
       "2       GUT_GENOME000008  MGYG-HGUT-00003    ERS3638033  Cultured (human gut)   \n",
       "3       GUT_GENOME000010  MGYG-HGUT-00004    ERS3638034  Cultured (human gut)   \n",
       "4       GUT_GENOME000017  MGYG-HGUT-00005    ERS3638035  Cultured (human gut)   \n",
       "\n",
       "   Reference genes  Core genes1  % Core genes1  Accessory genes1  \\\n",
       "0             3182          NaN            NaN               NaN   \n",
       "1             3920       1821.0          46.45           34621.0   \n",
       "2             2690       1522.0          56.58           19714.0   \n",
       "3             3634          NaN            NaN               NaN   \n",
       "4             3755          NaN            NaN               NaN   \n",
       "\n",
       "   Pan-genome size1  Number of SNVs  ...  Unnamed: 16  Unnamed: 17  \\\n",
       "0               NaN         16351.0  ...            0            1   \n",
       "1           36442.0        428560.0  ...           49          162   \n",
       "2           21236.0        516967.0  ...          244          676   \n",
       "3               NaN         58094.0  ...            5            5   \n",
       "4               NaN             NaN  ...            0            0   \n",
       "\n",
       "   Number of 97% ANI clusters  Number of 99% ANI clusters  \\\n",
       "0                           2                           4   \n",
       "1                         134                         327   \n",
       "2                          42                         561   \n",
       "3                           3                          17   \n",
       "4                           1                           2   \n",
       "\n",
       "   Number of genomes (total)  Number of genomes (non-redundant)  \\\n",
       "0                          4                                  3   \n",
       "1                        358                                308   \n",
       "2                       1178                               1000   \n",
       "3                         24                                 22   \n",
       "4                          2                                  2   \n",
       "\n",
       "   Number of genomes (near-complete)  Number of samples  \\\n",
       "0                                  1                  3   \n",
       "1                                120                308   \n",
       "2                                593               1003   \n",
       "3                                  8                 23   \n",
       "4                                  1                  2   \n",
       "\n",
       "                             Taxonomy lineage (GTDB)  \\\n",
       "0  d__Bacteria;p__Firmicutes_A;c__Clostridia;o__P...   \n",
       "1  d__Bacteria;p__Firmicutes_A;c__Clostridia;o__L...   \n",
       "2  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "3  d__Bacteria;p__Firmicutes_A;c__Clostridia;o__O...   \n",
       "4  d__Bacteria;p__Firmicutes_A;c__Clostridia;o__P...   \n",
       "\n",
       "                             Taxonomy lineage (NCBI)  \n",
       "0  d__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...  \n",
       "1  d__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...  \n",
       "2  d__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...  \n",
       "3  d__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...  \n",
       "4  d__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_data.sort_values('Number of genomes (near-complete)', ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_species = meta_data.iloc[:100, -1].apply(lambda x: ' '.join(x.split(';')[-2:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_species_row(meta_data, species_str):\n",
    "    mask = meta_data.iloc[:, -1].apply(lambda x: species_str in x)\n",
    "    return meta_data[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species representative</th>\n",
       "      <th>MGnify accession</th>\n",
       "      <th>ENA accession</th>\n",
       "      <th>Status</th>\n",
       "      <th>Reference genes</th>\n",
       "      <th>Core genes1</th>\n",
       "      <th>% Core genes1</th>\n",
       "      <th>Accessory genes1</th>\n",
       "      <th>Pan-genome size1</th>\n",
       "      <th>Number of SNVs</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 16</th>\n",
       "      <th>Unnamed: 17</th>\n",
       "      <th>Number of 97% ANI clusters</th>\n",
       "      <th>Number of 99% ANI clusters</th>\n",
       "      <th>Number of genomes (total)</th>\n",
       "      <th>Number of genomes (non-redundant)</th>\n",
       "      <th>Number of genomes (near-complete)</th>\n",
       "      <th>Number of samples</th>\n",
       "      <th>Taxonomy lineage (GTDB)</th>\n",
       "      <th>Taxonomy lineage (NCBI)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2477</th>\n",
       "      <td>GUT_GENOME143505</td>\n",
       "      <td>MGYG-HGUT-02478</td>\n",
       "      <td>ERS3640508</td>\n",
       "      <td>Cultured (human gut)</td>\n",
       "      <td>4517</td>\n",
       "      <td>2259.0</td>\n",
       "      <td>50.01</td>\n",
       "      <td>192070.0</td>\n",
       "      <td>194329.0</td>\n",
       "      <td>842418.0</td>\n",
       "      <td>...</td>\n",
       "      <td>992</td>\n",
       "      <td>2343</td>\n",
       "      <td>276</td>\n",
       "      <td>2853</td>\n",
       "      <td>5750</td>\n",
       "      <td>3649</td>\n",
       "      <td>2090</td>\n",
       "      <td>3685</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>d__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Species representative MGnify accession ENA accession  \\\n",
       "2477       GUT_GENOME143505  MGYG-HGUT-02478    ERS3640508   \n",
       "\n",
       "                    Status  Reference genes  Core genes1  % Core genes1  \\\n",
       "2477  Cultured (human gut)             4517       2259.0          50.01   \n",
       "\n",
       "      Accessory genes1  Pan-genome size1  Number of SNVs  ...  Unnamed: 16  \\\n",
       "2477          192070.0          194329.0        842418.0  ...          992   \n",
       "\n",
       "      Unnamed: 17  Number of 97% ANI clusters  Number of 99% ANI clusters  \\\n",
       "2477         2343                         276                        2853   \n",
       "\n",
       "      Number of genomes (total)  Number of genomes (non-redundant)  \\\n",
       "2477                       5750                               3649   \n",
       "\n",
       "      Number of genomes (near-complete)  Number of samples  \\\n",
       "2477                               2090               3685   \n",
       "\n",
       "                                Taxonomy lineage (GTDB)  \\\n",
       "2477  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "\n",
       "                                Taxonomy lineage (NCBI)  \n",
       "2477  d__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...  \n",
       "\n",
       "[1 rows x 26 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# finding B. vulgatus (B. dorei)\n",
    "find_species_row(meta_data, 'dorei')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "row = find_species_row(meta_data, 'massiliensis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2491    d__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...\n",
       "Name: Taxonomy lineage (NCBI), dtype: object"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row.iloc[:, -1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E. rectale's accession is MGYG-HGUT-02492\n"
     ]
    }
   ],
   "source": [
    "mgnify_accession = find_species_row(meta_data, 'rectale')['MGnify accession'].squeeze()\n",
    "print(\"E. rectale's accession is {}\".format(mgnify_accession))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def accession_to_species_name(accession):\n",
    "    row = meta_data[meta_data['MGnify accession']==accession]\n",
    "    return ';'.join(row['Taxonomy lineage (NCBI)'].item().split(';')[-2:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'g__Ruminococcus;s__'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "accession_to_species_name('MGYG-HGUT-02321')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Downloading ref genome data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The '.faa' file is \"All predicted CDS\"\n",
    "\n",
    "'fna' file is the \"DNA sequence FASTA file of the genome assembly of the species representative\"\n",
    "\n",
    "A better way to download might be using the API, then we don't need to deal with the ftp address directly:\n",
    "\"https://www.ebi.ac.uk/metagenomics/api/v1/genomes/MGYG000000001/downloads/MGYG000000001.faa\"\n",
    "\n",
    "This is the same as using this website by hand:\n",
    "https://www.ebi.ac.uk/metagenomics/genomes/MGYG000000001#downloads\n",
    "\n",
    "However, these addresses are v2.0, so they have slightly different genome IDs: \"MGYG-HGUT-00001\" vs \"MGYG000000001\", for this reason I think we should download everything from the ftp address."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgnify_accession = 'MGYG-HGUT-00001'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download using API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def accession_to_fileid(accession):\n",
    "    # reformatting the id\n",
    "    # my understanding is that we need to replace the HGUT part by 0000 and remove the -\n",
    "    items = accession.split('-')\n",
    "    return '0000'.join([items[0], items[-1]])\n",
    "\n",
    "fileid = accession_to_fileid(mgnify_accession)\n",
    "api_address = \"https://www.ebi.ac.uk/metagenomics/api/v1/genomes/{}/downloads/\".format(fileid)\n",
    "filename = \"{}.faa\".format(fileid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.ebi.ac.uk/metagenomics/api/v1/genomes/MGYG000000001/downloads/MGYG000000001.faa'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "api_address + filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "# download the sequence and the annotation\n",
    "filename = \"{}.faa\".format(fileid)\n",
    "r = requests.get(api_address + filename)\n",
    "open(filename, 'wb').write(r.content)\n",
    "\n",
    "filename = \"{}.fna\".format(fileid)\n",
    "r = requests.get(api_address + filename)\n",
    "open(filename, 'wb').write(r.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download using ftp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the ftp address has an upper level folder that omits the last two number of the accession id\n",
    "ftp_base = \"http://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/human-gut/v1.0/uhgg_catalogue/{}/{}/genome/\"\n",
    "ftp_address = ftp_base.format(mgnify_accession[:-2], mgnify_accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "http://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/human-gut/v1.0/uhgg_catalogue/MGYG-HGUT-000/MGYG-HGUT-00001/genome/\n"
     ]
    }
   ],
   "source": [
    "print(ftp_address)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = \"/Volumes/Botein/uhgg/reference_genomes/{}.fna\".format(mgnify_accession)\n",
    "r = requests.get(ftp_address + '{}.fna'.format(mgnify_accession))\n",
    "open(filename, 'wb').write(r.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now let's check the SNV table\n",
    "Let's verify that the reference genome makes sense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download the SNV table\n",
    "ftp_address = \"http://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/human-gut/v1.0/snv_catalogue/\"\n",
    "mgnify_accession = 'MGYG-HGUT-00001'\n",
    "\n",
    "filename = \"{}.tsv.tar.lz4\".format(mgnify_accession)\n",
    "r = requests.get(ftp_address + filename)\n",
    "open(filename, 'wb').write(r.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Decompress in bash using: \n",
    "```\n",
    "lz4 -dc < MGYG-HGUT-00001.tsv.tar.lz4 | tar xvf -\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "snv_df = pd.read_csv('MGYG-HGUT-00001_snvs.tsv', delimiter='\\t')  # using the file uploaded to github repo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>GUT_GENOME091053</th>\n",
       "      <th>GUT_GENOME178957</th>\n",
       "      <th>GUT_GENOME212267</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GUT_GENOME000001_106</td>\n",
       "      <td>1002</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GUT_GENOME000001_106</td>\n",
       "      <td>1020</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GUT_GENOME000001_106</td>\n",
       "      <td>1023</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GUT_GENOME000001_106</td>\n",
       "      <td>1211</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GUT_GENOME000001_106</td>\n",
       "      <td>1213</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GUT_GENOME000001_106</td>\n",
       "      <td>1217</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GUT_GENOME000001_106</td>\n",
       "      <td>1236</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GUT_GENOME000001_106</td>\n",
       "      <td>1263</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GUT_GENOME000001_106</td>\n",
       "      <td>1288</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GUT_GENOME000001_106</td>\n",
       "      <td>1302</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Contig   Pos Ref Alt  GUT_GENOME091053  GUT_GENOME178957  \\\n",
       "0  GUT_GENOME000001_106  1002   C   T                 1                 1   \n",
       "1  GUT_GENOME000001_106  1020   T   C                 1                 1   \n",
       "2  GUT_GENOME000001_106  1023   C   T                 1                 1   \n",
       "3  GUT_GENOME000001_106  1211   A   G                 1                 1   \n",
       "4  GUT_GENOME000001_106  1213   T   G                 1                 1   \n",
       "5  GUT_GENOME000001_106  1217   T   G                 1                 1   \n",
       "6  GUT_GENOME000001_106  1236   A   G                 1                 1   \n",
       "7  GUT_GENOME000001_106  1263   C   T                 1                 1   \n",
       "8  GUT_GENOME000001_106  1288   T   C                 1                 1   \n",
       "9  GUT_GENOME000001_106  1302   C   A                 1                 1   \n",
       "\n",
       "   GUT_GENOME212267  \n",
       "0                 1  \n",
       "1                 1  \n",
       "2                 1  \n",
       "3                 1  \n",
       "4                 1  \n",
       "5                 1  \n",
       "6                 1  \n",
       "7                 1  \n",
       "8                 1  \n",
       "9                 1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snv_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's check the first contig\n",
    "desired_contig = snv_df.iloc[0, 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# parsing the fasta file for ref seqeunce\n",
    "# input_file = \"{}.fna\".format(mgnify_accession)\n",
    "filename = \"/Volumes/Botein/uhgg/reference_genomes/{}.fna\".format(mgnify_accession)\n",
    "input_file = filename\n",
    "fasta_sequences = SeqIO.parse(open(input_file),'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "all_records = []\n",
    "for record in fasta_sequences:\n",
    "    if record.id==desired_contig:\n",
    "        desired_record = record\n",
    "    all_records.append(record)\n",
    "print(len(all_records))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "contigs = [record.id for record in records]\n",
    "seqs = [record.seq for record in records]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "contig_seq_map = dict(zip(contigs, seqs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCTAAAAATTTTATAATTTCGTTAGGCTTGTTTGCCATGCAAATTTTTAAAAATATGAAACTTAATTTAACCTATCGGTTTATTTATCAATTCGAGAATAATTCTCATCATATCTAATTAACAAAACCTAACTAAATATTTATTAATTCTAATCTTAAATCATTTAATATTGTCTTTGTTGCTCTTAATCTTATTTGCTGTCTATTTCCATTAAATACGTATCTCTTTACTATTATCTTTCCATTTATACAAATCCCAAAATATACCAAACCAACTGGTTTTTCTTCACTTCCACCTTCTGGTCCTGCTATTCCTGTTGTTGATAGCCCTATATTGGTATTATGCCTTTTAGCTATACCTTCTGCCATTTCTCTAGCTGTTTCCTCACTAACTGCTCCAAACTTATCTAGAGTTTCTTTTTTCACACCTAAAGATTTCATCTTAGCTTCATTGGAATAAGTTACACAACCTTCCATAAAAACAGATGAAATTCCTGGATAATTTATAAGGTGAGATGATACCATTCCTCCTGTACATGATTCTGCTACACTTATAGTTAAATTTTTATTAACTAAAATCTCAGCAACAGTATCTTCTAATGAAATATCACCCTCTGAATATATGTAACTTCCTACTCTATCTTGAATTTTATGTACAACTGGATTTATTAAATTAATAGCTTCTTCTTTACTAGTAGCTTTTGCTGTAATTCTTAAAGTTACTTCCGTGTCCTTTGCATATGGAGCTACAGTTGGATTTGATTGTTCTTTTATTATATCTATTATCTCTTCTTCTAAAGAAGATTCACCAATTCCATAGAATCTAAGAGTTTTAGAAACCAAAATTTCATTTGTTAAATTTTTTAAATATGGCTTAACATTTTCTTCGAACATTGGCTTCATTTCTCTTGGTGGGCCTGGTAATACTATTATTGTTTTTTTTCCTTTTCTTAAAATTGCTCCTGGTGCTGTACCATTGTTATTTTCAAGTATTATTGCATCCTTCGGAAAGTATGCTTGTTTCTTGTTATTCTCAGTTGGTACTCTTCCTAATTTATTAAAATATATTTCTATCTTATTCCATGATTTTTCATCAAGTACTAATTCTTGATTAAAAAACTTAGCTGCAGTTTCTTTCGTTAAATCATCTTTAGTTGGACCAAGTCCTCCTGTAGTTATTACCATATCACTTCTATTTAAACTTTCTTCAAAATACTTTAATAATCTTTCCTCATTATCTCCCACAGTTCCTTGATAATACACCTCGACTCCTAAATCAGCTAATTCTTTTGCCAAGTATTGCGAGTTTGTATTTACTATATCTCCTAAAAGAATTTCTGTTCCTACAGTTATAATTTCTGCTCTCATTGTCATTCACCCCTTCAATAATAATGTTTACCATATACATAATTTTGATGTATTTTTATATCTAAATATCTTTATATAAAACAAAAGGGCATATTTCGAATATATTTAGTTTATCTCAGAAGAAAGATAGAAAATATAAATTTTCTACCAGCTCCTCTAAATAAACTAACCTATCCGAAACAGCCCATTTTCATTGGAGGCGACACCCAGATTCGAACTGGGGATCAAAGAGTTGCAGTCCACTGCCTTACCACTTGGCTATGTCGCCTTATTGATTACAAGTAAATTATATATATCTTTGGTAATTTGTGCAACATAATTAATAAATATACTTAAAAATTTTATTTATAATCTCAACCACTCCACTCATAAACTTATAAAGTAGTGGTACTTTCAGTTCAAATATTAATTCAATATAATTAACTATATGAAAAAACATAGAAAATAAAATCGTCAATTTAATTTGAAGATTTAATAAATAGGTTTTGTTAATTAGATATGATGAGAATTATTCTCGAATTGATAAATAAACCGATAGGTTAAATTAAGTTTCATATTTTTAAAAATTTGCATGGCAAACAAGCCTAACGAAATTATAAAATTTTTAGG\n",
      "Contig length: 1976\n"
     ]
    }
   ],
   "source": [
    "print(desired_record.seq)\n",
    "print(\"Contig length: {}\".format(len(desired_record.seq)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloaded ref, SNV table ref\n",
      "('C', 'C')\n",
      "('T', 'T')\n",
      "('C', 'C')\n",
      "('A', 'A')\n",
      "('T', 'T')\n",
      "('T', 'T')\n",
      "('A', 'A')\n",
      "('C', 'C')\n",
      "('T', 'T')\n",
      "('C', 'C')\n",
      "('A', 'A')\n",
      "('C', 'C')\n",
      "('C', 'C')\n",
      "('A', 'A')\n",
      "('T', 'T')\n",
      "('C', 'C')\n",
      "('A', 'A')\n",
      "('T', 'T')\n",
      "('A', 'A')\n",
      "('A', 'A')\n"
     ]
    }
   ],
   "source": [
    "print(\"Downloaded ref, SNV table ref\")\n",
    "for _, row in snv_df.head(20).iterrows():\n",
    "    # remember to change contig position to str index by subtracting 1 !!!\n",
    "    # took me a while to realize this !!!\n",
    "    print(desired_record.seq[int(row['Pos'])-1], row['Ref'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check for E. rectale\n",
    "gene_id = 14\n",
    "snv_df = pd.read_csv('/Volumes/Botein/uhgg/{}/{}-{}.tsv'.format(mgnify_accession, contig, gene_id), delimiter='\\t', header=None)\n",
    "snv_df.columns = header.split('\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene_id in range(100):\n",
    "    snv_df = pd.read_csv('/Volumes/Botein/uhgg/{}/{}-{}.tsv'.format(mgnify_accession, contig, gene_id), delimiter='\\t', header=None)\n",
    "    snv_df.columns = header.split('\\t')\n",
    "    for _, row in snv_df.iterrows():\n",
    "        # remember to change contig position to str index by subtracting 1 !!!\n",
    "        # took me a while to realize this !!!\n",
    "        contig = row['Contig']\n",
    "        ref_allele = seq[int(row['Pos'])-1]\n",
    "        snv_table_allele = row['Ref']\n",
    "        if ref_allele != snv_table_allele:\n",
    "            print(row)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have checked that the reference sequence is indeed what's reported in the SNV table, the next step should be parsing the gene annotations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing the SNV tables\n",
    "First step: split the SNV table by gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('..')\n",
    "import UHGG_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "snvs_path = '/Volumes/Botein/uhgg/MGYG-HGUT-00001_snvs.tsv'\n",
    "grouped_snvs_base = '/Volumes/Botein/uhgg/MGYG-HGUT-00001/'  # make sure to start with an empty folder\n",
    "gene_df = UHGG_utils.gff_to_df('/Volumes/Botein/uhgg/MGYG-HGUT-00001.gff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "UHGG_utils.process_SNVs_table(snvs_path, gene_df, grouped_snvs_base)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Second step: process each individual genes and compute all pairwise counts\n",
    "I've done the above for the bigger table of E. rectale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "mgnify_accession = 'MGYG-HGUT-02492'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'UHGG_utils' from '../UHGG_utils.pyc'>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reload(UHGG_utils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Device6/.pyenv/versions/microbiome/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2714: DtypeWarning: Columns (15) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "snvs_path = '/Volumes/Botein/uhgg/MGYG-HGUT-02492_snvs.tsv'\n",
    "grouped_snvs_base = '/Volumes/Botein/uhgg/MGYG-HGUT-02492/'  # make sure to start with an empty folder\n",
    "gene_df = UHGG_utils.gff_to_df('/Volumes/Botein/uhgg/MGYG-HGUT-02492.gff')\n",
    "gene_df = gene_df.set_index('Gene ID')\n",
    "genomes_metadata = pd.read_csv('/Volumes/Botein/uhgg/genomes-nr_metadata.tsv', delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the header that contains all the sample names\n",
    "header = UHGG_utils.get_SNVs_table_header(snvs_path)\n",
    "genome_names = UHGG_utils.get_genome_names_from_table_header(header)\n",
    "genome_mask = UHGG_utils.get_non_redundant_genome_mask(genomes_metadata, mgnify_accession, genome_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'UHGG_utils' from '../UHGG_utils.py'>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reload(UHGG_utils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "core_genes = UHGG_utils.load_core_genes(mgnify_accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "core_genes = [int(x) for x in core_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished 0 at 21:50:12\n",
      "Finished 100 at 21:52:57\n",
      "Finished 200 at 21:54:29\n",
      "Finished 300 at 21:56:39\n",
      "Finished 400 at 22:01:37\n",
      "Finished 500 at 22:09:19\n",
      "Finished 600 at 22:17:17\n",
      "Finished 700 at 22:23:00\n",
      "Finished 800 at 22:28:55\n",
      "Finished 900 at 22:34:25\n",
      "Finished 1000 at 22:39:02\n",
      "Finished 1100 at 22:39:02\n",
      "Finished 1200 at 22:42:24\n",
      "Finished 1300 at 22:49:35\n",
      "Finished 1400 at 22:55:22\n",
      "Finished 1500 at 22:58:46\n",
      "Finished 1600 at 23:02:50\n",
      "Finished 1700 at 23:04:04\n",
      "Finished 1800 at 23:05:40\n",
      "Finished 1900 at 23:08:17\n",
      "Finished 2000 at 23:11:44\n",
      "Finished 2100 at 23:12:48\n",
      "Finished 2200 at 23:13:47\n",
      "Finished 2300 at 23:15:19\n",
      "Finished 2400 at 23:16:19\n",
      "Finished 2500 at 23:16:19\n",
      "Finished 2600 at 23:16:31\n",
      "Finished 2700 at 23:16:31\n",
      "Finished 2800 at 23:19:04\n",
      "Finished 2900 at 23:24:44\n",
      "Finished 3000 at 23:25:11\n",
      "Finished 3100 at 23:29:29\n",
      "Finished 3200 at 23:30:29\n",
      "Finished 3300 at 23:32:22\n",
      "Finished 3400 at 23:36:23\n",
      "Finished 3429 at 23:38:44\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "gene_files = os.listdir(grouped_snvs_base)\n",
    "\n",
    "output_path = '/Volumes/Botein/uhgg/E_rectale_site_pairs_083022.txt'\n",
    "with open(output_path, 'w') as f:\n",
    "    f.write(' '.join(('n11', 'n10', 'n01', 'n00', 'ell')) + '\\n')\n",
    "\n",
    "processed = 0\n",
    "\n",
    "for gene_file in gene_files:\n",
    "    gene_file_path = grouped_snvs_base + gene_file\n",
    "    items = gene_file.split('.')[0].split('-')\n",
    "    gene_id = int(items[-1])\n",
    "    contig = items[0]\n",
    "    if processed % 100 == 0:\n",
    "        now = datetime.now()\n",
    "        current_time = now.strftime(\"%H:%M:%S\")\n",
    "        print(\"Finished {} at {}\".format(processed, current_time))\n",
    "    processed += 1\n",
    "    if gene_df.loc[gene_id, 'Type'] != 'CDS':\n",
    "        continue\n",
    "    elif gene_id not in core_genes:\n",
    "        continue\n",
    "    UHGG_utils.process_SNV_table_single_gene(mgnify_accession, header, genomes_metadata, gene_file_path, contig, gene_id, output_path)\n",
    "\n",
    "now = datetime.now()\n",
    "current_time = now.strftime(\"%H:%M:%S\")\n",
    "print(\"Finished {} at {}\".format(processed, current_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'UHGG_utils' from '../UHGG_utils.py'>"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reload(UHGG_utils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished 0 at 11:09:36\n",
      "Finished 100 at 11:18:08\n",
      "Finished 200 at 11:27:33\n",
      "Finished 300 at 11:37:04\n",
      "Finished 400 at 11:46:45\n",
      "Finished 500 at 11:56:57\n",
      "Finished 600 at 12:07:07\n",
      "Finished 700 at 12:15:20\n",
      "Finished 800 at 12:23:30\n",
      "Finished 900 at 12:31:54\n",
      "Finished 999 at 12:40:47\n"
     ]
    }
   ],
   "source": [
    "# now process pairs of random genes\n",
    "gene_files = os.listdir(grouped_snvs_base)\n",
    "\n",
    "output_path = '/Volumes/Botein/uhgg/E_rectale_site_pairs_083022.txt'\n",
    "n_pairs = 1000\n",
    "for i in range(n_pairs):\n",
    "    gene1, gene2 = UHGG_utils.sample_random_pair_of_genes(gene_files, core_genes, gene_df)\n",
    "    gene_file_path1 = grouped_snvs_base + gene1\n",
    "    gene_file_path2 = grouped_snvs_base + gene2\n",
    "    if i % 100 == 0:\n",
    "        now = datetime.now()\n",
    "        current_time = now.strftime(\"%H:%M:%S\")\n",
    "        print(\"Finished {} at {}\".format(i, current_time))\n",
    "    UHGG_utils.process_SNV_table_between_genes(gene_file_path1, gene_file_path2, genome_mask, output_path)\n",
    "\n",
    "now = datetime.now()\n",
    "current_time = now.strftime(\"%H:%M:%S\")\n",
    "print(\"Finished {} at {}\".format(i, current_time))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the below codes to load the final results and resave into a numpy binary format for faster future loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# very slow to parse the csv using np; slightly faster if using pandas; best to resave into binary format like npy!\n",
    "output_path = '/Volumes/Botein/uhgg/E_rectale_site_pairs_083022.txt'\n",
    "res = pd.read_csv(output_path, delimiter=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = res.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('/Volumes/Botein/uhgg/E_rectale_site_pairs_083022', res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.32 ms, sys: 6.52 s, total: 6.52 s\n",
      "Wall time: 10.8 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "res = np.load('/Volumes/Botein/uhgg/E_rectale_site_pairs.npy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Converting Ben's data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading data...\n",
      "CPU times: user 2min 30s, sys: 5.71 s, total: 2min 36s\n",
      "Wall time: 2min 44s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import gzip\n",
    "file = gzip.GzipFile('../erectale_snv_pairs.txt.gz',\"r\")\n",
    "\n",
    "\n",
    "file.readline() # header\n",
    "\n",
    "n11s = []\n",
    "n10s = []\n",
    "n01s = []\n",
    "n00s = []\n",
    "ells = []\n",
    "print \"Loading data...\"\n",
    "for line in file:\n",
    "    \n",
    "    items = line.split()\n",
    "    n11 = long(items[0])\n",
    "    n10 = long(items[1])\n",
    "    n01 = long(items[2])\n",
    "    n00 = long(items[3])\n",
    "    ell = long(items[4])\n",
    "\n",
    "    n11s.append(n11)\n",
    "    n10s.append(n10)\n",
    "    n01s.append(n01)\n",
    "    n00s.append(n00)\n",
    "    ells.append(ell)\n",
    "\n",
    "file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.sizeof(n11s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "n11s =numpy.array(n11s)*1.0\n",
    "n10s = numpy.array(n10s)*1.0\n",
    "n01s = numpy.array(n01s)*1.0\n",
    "n00s = numpy.array(n00s)*1.0\n",
    "ntots = n11s+n10s+n01s+n00s\n",
    "ells = numpy.array(ells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ns = np.concatenate([n11s, n10s, n01s, n00s, ells])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scratch space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_files = os.listdir(grouped_snvs_base)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'snvs_path' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-53-9fa6bb609390>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mmgnify_accession\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'MGYG-HGUT-02492'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mspecies_accession\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmgnify_accession\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mheader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mUHGG_utils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_SNVs_table_header\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnvs_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mgenome_names\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mgenome_mask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mUHGG_utils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_non_redundant_genome_mask\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenomes_metadata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mspecies_accession\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenome_names\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'snvs_path' is not defined"
     ]
    }
   ],
   "source": [
    "mgnify_accession = 'MGYG-HGUT-02492'\n",
    "species_accession = mgnify_accession\n",
    "header = UHGG_utils.get_SNVs_table_header(snvs_path)\n",
    "genome_names = header.split('\\t')[4:]\n",
    "genome_mask = UHGG_utils.get_non_redundant_genome_mask(genomes_metadata, species_accession, genome_names)\n",
    "num_genomes = genome_mask.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "cumu_pair_snvs = np.zeros((num_genomes, num_genomes))\n",
    "cumu_pair_passed = np.zeros((num_genomes, num_genomes))\n",
    "cumu_diff_genes = np.zeros((num_genomes, num_genomes))\n",
    "cumu_passed_genes = np.zeros((num_genomes, num_genomes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished 0 at 13:58:12\n",
      "Finished 10 at 14:00:29\n",
      "Finished 20 at 14:02:55\n",
      "Finished 30 at 14:05:15\n",
      "Finished 40 at 14:07:00\n",
      "Finished 50 at 14:09:20\n",
      "Finished 60 at 14:11:40\n",
      "Finished 70 at 14:14:44\n",
      "Finished 80 at 14:17:33\n",
      "Finished 90 at 14:20:05\n",
      "Finished 100 at 14:22:26\n",
      "Finished 110 at 14:24:28\n",
      "Finished 120 at 14:26:40\n",
      "Finished 130 at 14:28:42\n",
      "Finished 140 at 14:30:31\n",
      "Finished 150 at 14:32:47\n",
      "Finished 160 at 14:35:27\n",
      "Finished 170 at 14:38:45\n",
      "Finished 180 at 14:42:35\n",
      "Finished 190 at 14:45:52\n",
      "Finished 200 at 14:48:19\n",
      "Finished 210 at 14:50:43\n",
      "Finished 220 at 14:53:09\n",
      "Finished 230 at 14:55:04\n",
      "Finished 240 at 14:57:18\n",
      "Finished 250 at 14:59:25\n",
      "Finished 260 at 15:01:34\n",
      "Finished 270 at 15:03:52\n",
      "Finished 280 at 15:05:26\n",
      "Finished 290 at 15:07:41\n",
      "Finished 300 at 15:10:28\n",
      "Finished 310 at 15:12:50\n",
      "Finished 320 at 15:15:12\n",
      "Finished 330 at 15:18:17\n",
      "Finished 340 at 15:21:06\n",
      "Finished 350 at 15:23:16\n",
      "Finished 360 at 15:26:11\n",
      "Finished 370 at 15:28:02\n",
      "Finished 380 at 15:30:44\n",
      "Finished 390 at 15:33:05\n",
      "Finished 400 at 15:35:33\n",
      "Finished 410 at 15:37:33\n",
      "Finished 420 at 15:40:01\n",
      "Finished 430 at 15:42:01\n",
      "Finished 440 at 15:44:52\n",
      "Finished 450 at 15:47:05\n",
      "Finished 460 at 15:49:51\n",
      "Finished 470 at 15:52:42\n",
      "Finished 480 at 15:54:44\n",
      "Finished 490 at 15:56:59\n",
      "Finished 500 at 15:59:04\n",
      "Finished 510 at 16:00:49\n",
      "Finished 520 at 16:03:04\n",
      "Finished 530 at 16:05:26\n",
      "Finished 540 at 16:08:21\n",
      "Finished 550 at 16:11:03\n",
      "Finished 560 at 16:13:48\n",
      "Finished 570 at 16:16:11\n",
      "Finished 580 at 16:18:38\n",
      "Finished 590 at 16:21:50\n",
      "Finished 600 at 16:23:58\n",
      "Finished 610 at 16:26:25\n",
      "Finished 620 at 16:28:46\n",
      "Finished 630 at 16:31:35\n",
      "Finished 640 at 16:33:48\n",
      "Finished 650 at 16:36:44\n",
      "Finished 660 at 16:39:06\n",
      "Finished 670 at 16:41:19\n",
      "Finished 680 at 16:43:26\n",
      "Finished 690 at 16:46:00\n",
      "Finished 700 at 16:48:05\n",
      "Finished 710 at 16:50:39\n",
      "Finished 720 at 16:53:13\n",
      "Finished 730 at 16:56:22\n",
      "Finished 740 at 16:59:04\n",
      "Finished 750 at 17:01:40\n",
      "Finished 760 at 17:04:37\n",
      "Finished 770 at 17:07:05\n",
      "Finished 780 at 17:08:59\n",
      "Finished 790 at 17:11:59\n",
      "Finished 800 at 17:14:40\n",
      "Finished 810 at 17:16:46\n",
      "Finished 820 at 17:18:32\n",
      "Finished 830 at 17:20:19\n",
      "Finished 840 at 17:22:22\n",
      "Finished 850 at 17:24:06\n",
      "Finished 860 at 17:25:47\n",
      "Finished 870 at 17:28:16\n",
      "Finished 880 at 17:30:40\n",
      "Finished 890 at 17:32:53\n",
      "Finished 900 at 17:35:19\n",
      "Finished 910 at 17:37:11\n",
      "Finished 920 at 17:39:36\n",
      "Finished 930 at 17:42:27\n",
      "Finished 940 at 17:44:03\n",
      "Finished 950 at 17:46:28\n",
      "Finished 960 at 17:48:32\n",
      "Finished 970 at 17:50:16\n",
      "Finished 980 at 17:52:08\n",
      "Finished 990 at 17:54:02\n",
      "Finished 1000 at 17:55:55\n",
      "Finished 1010 at 17:57:41\n",
      "Finished 1020 at 18:00:12\n",
      "Finished 1030 at 18:01:56\n",
      "Finished 1040 at 18:03:53\n",
      "Finished 1050 at 18:05:58\n",
      "Finished 1060 at 18:07:57\n",
      "Finished 1070 at 18:09:47\n",
      "Finished 1080 at 18:11:17\n",
      "Finished 1090 at 18:13:48\n",
      "Finished 1100 at 18:15:53\n",
      "Finished 1110 at 18:19:20\n"
     ]
    }
   ],
   "source": [
    "# script to compute pairwise distance\n",
    "from datetime import datetime\n",
    "processed = 0\n",
    "total_sites = 0\n",
    "\n",
    "for gene_file in gene_files:\n",
    "    gene_file_path = grouped_snvs_base + gene_file\n",
    "    items = gene_file.split('.')[0].split('-')\n",
    "    gene_id = int(items[-1])  # this gene id is assigned by us according to the reference genome's annotation order\n",
    "    contig = items[0]\n",
    "    if gene_df.loc[gene_id, 'Type'] != 'CDS':\n",
    "        continue\n",
    "    elif str(gene_id) not in species_core_genes:\n",
    "        # check if gene is a \"core gene\"\n",
    "        continue\n",
    "    \n",
    "    # load SNV table\n",
    "    dat = pd.read_csv(gene_file_path, delimiter='\\t', header=None)\n",
    "    dat.columns = header.split('\\t')\n",
    "    dat.sort_values('Pos', inplace=True)\n",
    "\n",
    "    # filtering sites with more than two alleles\n",
    "    biallelic_dat = dat.groupby('Pos').filter(lambda x: x.shape[0] == 1)\n",
    "\n",
    "    # converting to numpy array for faster arithmetics\n",
    "    snvs = biallelic_dat.iloc[:, 4:].astype(int).to_numpy()\n",
    "    snvs = snvs[:, genome_mask]\n",
    "\n",
    "    zeros = (snvs == 0).astype(int)\n",
    "    ones = (snvs == 1).astype(int)\n",
    "    covered = (snvs != 255).astype(int)\n",
    "    \n",
    "    pair_snvs = np.dot(zeros.T, ones)  # using dot product to count the number of sites that are different\n",
    "    pair_snvs = pair_snvs + pair_snvs.T\n",
    "    pair_passed = np.dot(covered.T, covered)\n",
    "    \n",
    "    total_sites += snvs.shape[0]\n",
    "    cumu_pair_snvs += pair_snvs\n",
    "    cumu_pair_passed += pair_passed  # record the number of covered sites between a pair\n",
    "    cumu_diff_genes += pair_snvs > 0  # record if a pair has non identical genes\n",
    "    cumu_passed_genes += pair_passed > 0 # record if a gene has any covered sites between a pair\n",
    "    \n",
    "    if processed % 10 == 0:\n",
    "        now = datetime.now()\n",
    "        current_time = now.strftime(\"%H:%M:%S\")\n",
    "        print(\"Finished {} at {}\".format(processed, current_time))\n",
    "    processed += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY0AAAD8CAYAAACLrvgBAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi40LCBodHRwOi8vbWF0cGxvdGxpYi5vcmcv7US4rQAAERtJREFUeJzt3X+snmV9x/H3Z0UQNAJCR7RlOzV2WyqbETtgcb8CE4oslmRq0G00htg/xE3HnBRdgtOwgDFDyZSlGUxIzIAxF5pRaRhItpmBtP4ACzKOgNAOpRaEKRGEfffHuaAPzTmnV3t+POfH+5U86f187+t+ruu5c04/57rv+7mfVBWSJPX4uWEPQJI0fxgakqRuhoYkqZuhIUnqZmhIkroZGpKkboaGJKmboSFJ6mZoSJK6HTTsAUy3o48+ukZGRoY9DEmaV7Zt2/bDqlq6r3YLLjRGRkbYunXrsIchSfNKku/1tPPwlCSpm6EhSepmaEiSuhkakqRuhoYkqZuhIUnqZmhIkroZGpKkboaGJKnbgvtEuOafkQ03vrj80MVnDHEkkvbFmYYkqZuhIUnqZmhIkroZGpKkboaGJKmboSFJ6mZoSJK6GRqSpG6GhiSpm6EhSepmaEiSuhkakqRu3rBQQzF4k0JJ84czDUlSN0NDktTN0JAkdTM0JEndDA1JUreu0EjyZ0m2J/l2kn9M8vIkK5LckWQ0ybVJDm5tD2nPR9v6kYHXuaDV70ty2kB9TauNJtkwUB+3D0nScOwzNJIsA/4UWF1VxwFLgLOAS4BLq+r1wBPAOW2Tc4AnWv3S1o4kq9p2bwDWAJ9PsiTJEuBzwOnAKuDdrS2T9CFJGoLew1MHAYcmOQg4DHgUOBm4vq2/CjizLa9tz2nrT0mSVr+mqp6pqgeBUeCE9hitqgeq6lngGmBt22aiPqQDNrLhxhcfkvbPPkOjqnYCnwYeZiwsngS2AT+qqudasx3Asra8DHikbftca3/UYH2vbSaqHzVJH5KkIeg5PHUkY7OEFcBrgVcwdnhpzkiyPsnWJFt37do17OFI0oLVc3jq94AHq2pXVf0M+BLwFuCIdrgKYDmwsy3vBI4FaOsPB3YP1vfaZqL67kn6eImq2lhVq6tq9dKlSzvekiTpQPSExsPASUkOa+cZTgHuAb4CvKO1WQfc0JY3tee09bdWVbX6We3qqhXASuBrwJ3Aynal1MGMnSzf1LaZqA9J0hD0nNO4g7GT0V8H7m7bbATOB85LMsrY+Ycr2iZXAEe1+nnAhvY624HrGAucm4Bzq+r5ds7iA8AW4F7gutaWSfqQJA1B111uq+pC4MK9yg8wduXT3m1/Crxzgte5CLhonPpmYPM49XH7kCQNh58IlyR1MzQkSd0MDUlSN7+5T4va4KfCH7r4jCGORJofnGlIkroZGpKkboaGJKmboSFJ6mZoSJK6GRqSpG6GhiSpm6EhSepmaEiSuhkakqRuhoYkqZuhIUnqZmhIkroZGpKkboaGJKmboSFJ6mZoSJK6GRqSpG6GhiSpm98RrjnF7+yW5jZnGpKkboaGJKmboSFJ6mZoSJK6GRqSpG6GhiSpm6EhSepmaEiSuhkakqRuhoYkqZuhIUnqZmhIkroZGpKkbl2hkeSIJNcn+U6Se5P8RpJXJ7k5yf3t3yNb2yS5LMlokruSHD/wOuta+/uTrBuovznJ3W2by5Kk1cftQ5I0HL0zjc8CN1XVrwBvBO4FNgC3VNVK4Jb2HOB0YGV7rAcuh7EAAC4ETgROAC4cCIHLgfcNbLem1SfqQ5I0BPsMjSSHA78NXAFQVc9W1Y+AtcBVrdlVwJlteS1wdY25HTgiyWuA04Cbq+rxqnoCuBlY09a9qqpur6oCrt7rtcbrQ5I0BD1fwrQC2AX8Q5I3AtuADwLHVNWjrc33gWPa8jLgkYHtd7TaZPUd49SZpA9pvwx+uZOkA9dzeOog4Hjg8qp6E/AT9jpM1GYINf3D6+sjyfokW5Ns3bVr10wOQ5IWtZ7Q2AHsqKo72vPrGQuRH7RDS7R/H2vrdwLHDmy/vNUmqy8fp84kfbxEVW2sqtVVtXrp0qUdb0mSdCD2GRpV9X3gkSS/3EqnAPcAm4AXroBaB9zQljcBZ7erqE4CnmyHmLYApyY5sp0APxXY0tY9leSkdtXU2Xu91nh9SJKGoOecBsCfAF9McjDwAPBexgLnuiTnAN8D3tXabgbeBowCT7e2VNXjST4J3NnafaKqHm/L7we+ABwKfLk9AC6eoA9J0hB0hUZVfRNYPc6qU8ZpW8C5E7zOlcCV49S3AseNU989Xh+SpOHwE+GSpG6GhiSpm6EhSepmaEiSuhkakqRuhoYkqZuhIUnqZmhIkroZGpKkboaGJKmboSFJ6mZoSJK6GRqSpG6GhiSpm6EhSepmaEiSuhkakqRuvV/3Kk3ZyIYbhz0ESVPkTEOS1M3QkCR1MzQkSd0MDUlSN0NDktTN0JAkdfOSW81Ze1+i+9DFZwxpJJJe4ExDktTN0JAkdTM0JEndPKehBcvblkjTz5mGJKmbMw2pGZyZeKWWND5nGpKkboaGJKmboSFJ6mZoSJK6GRqSpG6GhiSpW3doJFmS5BtJ/rU9X5HkjiSjSa5NcnCrH9Kej7b1IwOvcUGr35fktIH6mlYbTbJhoD5uH5Kk4difmcYHgXsHnl8CXFpVrweeAM5p9XOAJ1r90taOJKuAs4A3AGuAz7cgWgJ8DjgdWAW8u7WdrA9J0hB0hUaS5cAZwN+35wFOBq5vTa4CzmzLa9tz2vpTWvu1wDVV9UxVPQiMAie0x2hVPVBVzwLXAGv30YckaQh6ZxqfAT4C/F97fhTwo6p6rj3fASxry8uARwDa+idb+xfre20zUX2yPiRJQ7DP0Ejy+8BjVbVtFsZzQJKsT7I1ydZdu3YNeziStGD1zDTeArw9yUOMHTo6GfgscESSF+5dtRzY2ZZ3AscCtPWHA7sH63ttM1F99yR9vERVbayq1VW1eunSpR1vSZJ0IPYZGlV1QVUtr6oRxk5k31pVfwh8BXhHa7YOuKEtb2rPaetvrapq9bPa1VUrgJXA14A7gZXtSqmDWx+b2jYT9SFJGoKpfE7jfOC8JKOMnX+4otWvAI5q9fOADQBVtR24DrgHuAk4t6qeb+csPgBsYezqrOta28n6kCQNwX7dGr2qbgNua8sPMHbl095tfgq8c4LtLwIuGqe+Gdg8Tn3cPqSJ+MVL0szyE+GSpG5+CZPmDb8kSRo+ZxqSpG6GhiSpm6EhSepmaEiSuhkakqRuhoYkqZuhIUnqZmhIkrr54T7NS94uRBoOZxqSpG7ONKR5zturaDY505AkdTM0JEndDA1JUjdDQ5LUzdCQJHUzNCRJ3QwNSVI3Q0OS1M3QkCR18xPh0jzjfbc0TM40JEndnGlIC4j3odJMc6YhSermTEMah3+xS+NzpiFJ6mZoSJK6GRqSpG6GhiSpm6EhSepmaEiSunnJrTQPeOsQzRXONCRJ3QwNSVI3Q0OS1G2foZHk2CRfSXJPku1JPtjqr05yc5L7279HtnqSXJZkNMldSY4feK11rf39SdYN1N+c5O62zWVJMlkfkqTh6JlpPAf8eVWtAk4Czk2yCtgA3FJVK4Fb2nOA04GV7bEeuBzGAgC4EDgROAG4cCAELgfeN7DdmlafqA9J0hDs8+qpqnoUeLQt/2+Se4FlwFrgd1uzq4DbgPNb/eqqKuD2JEckeU1re3NVPQ6Q5GZgTZLbgFdV1e2tfjVwJvDlSfqQFjyvmNJctF+X3CYZAd4E3AEc0wIF4PvAMW15GfDIwGY7Wm2y+o5x6kzShzRrvOOttEf3ifAkrwT+GfhQVT01uK7NKmqax/YSk/WRZH2SrUm27tq1ayaHIUmLWtdMI8nLGAuML1bVl1r5B0leU1WPtsNPj7X6TuDYgc2Xt9pO9hxqeqF+W6svH6f9ZH28RFVtBDYCrF69ekbDS/vHQyzSwtJz9VSAK4B7q+pvBlZtAl64AmodcMNA/ex2FdVJwJPtENMW4NQkR7YT4KcCW9q6p5Kc1Po6e6/XGq8PSdIQ9Mw03gL8MXB3km+22keBi4HrkpwDfA94V1u3GXgbMAo8DbwXoKoeT/JJ4M7W7hMvnBQH3g98ATiUsRPgX271ifqQJA1Bz9VT/wlkgtWnjNO+gHMneK0rgSvHqW8Fjhunvnu8PiRJw+ENC6U5xHNAmusMDWmB8lJhzQTvPSVJ6uZMQ9oP/vWuxc6ZhiSpm6EhSepmaEiSuhkakqRungiXDpAnxbUYGRrSkM3GB/oMOE0XD09Jkro505CGwNuFaL4yNKRp0HP4Z64EhYeqNBUenpIkdTM0JEndPDwlzaC5ckhKmi6GhjTNDAotZB6ekiR1c6YhLWJeSaX95UxDktTN0JAkdTM0JEndDA1JUjdDQ5LUzdCQJHUzNCRJ3fychqadn4iWFi5nGpKkboaGJKmboSFJ6mZoSJK6eSJcEuDNC9XHmYYkqZuhIUnqZmhIkrp5TkPTwg/0SYuDMw1JUrc5HxpJ1iS5L8lokg3DHo8kLWZz+vBUkiXA54C3AjuAO5Nsqqp7hjsygYekpMVoTocGcAIwWlUPACS5BlgLGBpDYlAsDn5mQxOZ66GxDHhk4PkO4MQhjWVe8D91TbepBMh0/TxO1K/hNvvmemh0SbIeWN+e/jjJfQf4UkcDP5yeUc177os93BdNLhnOvsgl09Nmmi20n4tf7Gk010NjJ3DswPPlrfYSVbUR2DjVzpJsrarVU32dhcB9sYf7Yg/3xR6LdV/M9aun7gRWJlmR5GDgLGDTkMckSYvWnJ5pVNVzST4AbAGWAFdW1fYhD0uSFq05HRoAVbUZ2DxL3U35ENcC4r7Yw32xh/tij0W5L1JVwx6DJGmemOvnNCRJc8iiCY193Y4kySFJrm3r70gyMrDugla/L8lpsznumXCg+yLJW5NsS3J3+/fk2R77dJvKz0Vb/wtJfpzkw7M15pkyxd+RX0vyX0m2t5+Pl8/m2KfbFH5HXpbkqrYP7k1ywWyPfcZV1YJ/MHYS/bvA64CDgW8Bq/Zq837g79ryWcC1bXlVa38IsKK9zpJhv6ch7Ys3Aa9ty8cBO4f9foa1LwbWXw/8E/DhYb+fIf5cHATcBbyxPT9qEf+OvAe4pi0fBjwEjAz7PU3nY7HMNF68HUlVPQu8cDuSQWuBq9ry9cApSdLq11TVM1X1IDDaXm++OuB9UVXfqKr/afXtwKFJDpmVUc+MqfxckORM4EHG9sV8N5V9cSpwV1V9C6CqdlfV87M07pkwlX1RwCuSHAQcCjwLPDU7w54diyU0xrsdybKJ2lTVc8CTjP3F1LPtfDKVfTHoD4CvV9UzMzTO2XDA+yLJK4Hzgb+ahXHOhqn8XPwSUEm2JPl6ko/Mwnhn0lT2xfXAT4BHgYeBT1fV4zM94Nk05y+51dyT5A3AJYz9hblYfRy4tKp+3CYei9lBwG8Cvw48DdySZFtV3TLcYQ3FCcDzwGuBI4H/SPJv1W66uhAslplGz+1IXmzTppaHA7s7t51PprIvSLIc+Bfg7Kr67oyPdmZNZV+cCHwqyUPAh4CPtg+izldT2Rc7gH+vqh9W1dOMfa7q+Bkf8cyZyr54D3BTVf2sqh4DvgosqFuNLJbQ6LkdySZgXVt+B3BrjZ3N2gSc1a6WWAGsBL42S+OeCQe8L5IcAdwIbKiqr87aiGfOAe+LqvqtqhqpqhHgM8BfV9XfztbAZ8BUfke2AL+a5LD2H+jvML+/vmAq++Jh4GSAJK8ATgK+Myujni3DPhM/Ww/gbcB/M3ZVxMda7RPA29vyyxm7CmaUsVB43cC2H2vb3QecPuz3Mqx9AfwlY8drvznw+Plhv59h/VwMvMbHmedXT011XwB/xNgFAd8GPjXs9zKsfQG8stW3MxacfzHs9zLdDz8RLknqtlgOT0mSpoGhIUnqZmhIkroZGpKkboaGJKmboSFJ6mZoSJK6GRqSpG7/D4hVEtbex37HAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "uptri = np.triu_indices(cumu_diff_genes.shape[0], k=1)\n",
    "divs = cumu_pair_snvs[uptri] / cumu_pair_passed[uptri]\n",
    "_ = plt.hist(divs, bins=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi40LCBodHRwOi8vbWF0cGxvdGxpYi5vcmcv7US4rQAAD21JREFUeJzt3X+Ipdddx/H3p1u3xTbdFncF2R9NJJvQUP+wXNKKoJE2dtt0s1JL2WDRSsjSyhZRkQYUqvafFH+ANcF2NMs2BbPGUsqs3bJCbVjRpOzEYskmRNY1NhOF3Zo6IFXTtF//mBt7u2Zmnpn7a+657xcE7nPuc+/zPZnZ75z7PeeeJ1WFJKldL5t2AJKk8TLRS1LjTPSS1DgTvSQ1zkQvSY0z0UtS40z0ktQ4E70kNc5EL0mNe/k0L57kMHD4mmuuueuGG26YZiiSNHMee+yxr1fVno3Oy3bYAqHX69XS0tK0w5CkmZLksarqbXSepRtJapyJXpIaZ6KXpMZNNdEnOZxkYWVlZZphSFLTpproq+p0VR3btWvXNMOQpKZZupGkxpnoJalxJnpJatxUvxkrSS259u7P/9/jp++5bYqRfC9H9JLUuG2x1831118/zTAkacsGR/HblcsrJalxlm4kqXEmeklqnIlekhpnopekxpnoJalxJnpJapzfjJWkNaz1TddZWDs/aOSJPsnLgI8CrwGWqupTo76GJI3LrCXxLjqVbpKcSHI5yeNXtR9K8lSSi0nu7jcfAfYB3wKWRxuuJGmzuo7oTwL3Ag+82JBkB3AfcCurCf18kkXgRuDvquqTST4DfHGkEUvSFMzySL/TiL6qzgHPXdV8M3Cxqi5V1fPAKVZH88vAN/rnfHut90xyLMlSkqUrV65sPnJJUifDrLrZCzwzcLzcb/ss8PYkfwScW+vFVbVQVb2q6u3Zs2eIMCRJ6xn5ZGxVfRO4s8u57l4pSeM3TKJ/Ftg/cLyv39ZZVZ0GTvd6vbuGiEOShjLL9fcuhindnAcOJrkuyU7gKLC4mTdIcjjJwsrKyhBhSJLW02lEn+RB4BZgd5Jl4CNVdX+S48BZYAdwoqoubObijuglterqTwnTvLVgp0RfVXes0X4GODPSiCRJIzXVvW4s3UjS+HkrQUlqnCN6SWqcI3pJapz70UtS4yzdSFLjLN1IUuO8w5SkudT6tgeDrNFLUuOs0UtS46zRS1LjLN1IUuNM9JLUOBO9JDXOyVhJapyTsZLUOEs3ktQ4E70kNc4tECTNjXna9mCQI3pJapyJXpIa5/JKSWqcyyslqXGWbiSpcSZ6SWqciV6SGmeil6TGmeglqXEmeklq3Mi3QEhyC/BR4AJwqqoeHvU1JKmred32YFCnEX2SE0kuJ3n8qvZDSZ5KcjHJ3f3mAv4TeCWwPNpwJUmb1bV0cxI4NNiQZAdwH/AO4CbgjiQ3AX9TVe8APgz89uhClSRtRadEX1XngOeuar4ZuFhVl6rqeeAUcKSqvtN//hvAK0YWqSRpS4ap0e8Fnhk4XgbenOTdwNuB1wL3rvXiJMeAYwAHDhwYIgxJ0npGPhlbVZ8FPtvhvAVgAaDX69Wo45AkrRpmeeWzwP6B4339ts7cvVKSxm+YRH8eOJjkuiQ7gaPA4mbewN0rJWn8ui6vfBB4BLgxyXKSO6vqBeA4cBZ4Enioqi5s5uKO6CVp/DrV6KvqjjXazwBntnrxqjoNnO71endt9T0kSevzDlOS1DjvMCVJjRv58kpJmjb3t/lelm4kqXGWbiSpce5HL0mNs3QjSY2zdCNJjbN0I0mNM9FLUuOs0UtS46zRS1LjLN1IUuPcAkFSE9z2YG0mekmagME/RE/fc9tEr+1krCQ1zslYSWqck7GS1DgTvSQ1zkQvSY0z0UtS40z0ktQ4l1dKUuOm+oWpqjoNnO71endNMw5Js8lvw3Zj6UaSGmeil6TGmeglqXEmeklqnIlekhpnopekxo0l0Sd5VZKlJO8ax/tLkrrrlOiTnEhyOcnjV7UfSvJUkotJ7h546sPAQ6MMVJK0NV2/MHUSuBd44MWGJDuA+4BbgWXgfJJFYC/wBPDKkUYqSfglqa3olOir6lySa69qvhm4WFWXAJKcAo4ArwZeBdwE/FeSM1X1nZFFLEnalGG2QNgLPDNwvAy8uaqOAyR5P/D1tZJ8kmPAMYADBw4MEYYkaT1jW3VTVSer6i/XeX6hqnpV1duzZ8+4wpCkuTdMon8W2D9wvK/f1pm7V0rS+A2T6M8DB5Ncl2QncBRY3MwbeHNwSRq/rssrHwQeAW5Mspzkzqp6ATgOnAWeBB6qqgububgjekkav1TVtGOg1+vV0tLStMOQtE21tqTy6XtuG8n7JHmsqnobnecdpiSpcVNN9NboJWn83NRMkhpn6UaSGufNwSVtS61NwE6TpRtJapylG0lqnKtuJKlxlm4kqXFTnYyVpEFOwI6HNXpJapw1eklqnDV6SWqcNXpJE2ctfrJM9JImwuQ+PVNN9EkOA4evv/76aYYhaUxM7tuDe91IGimT+/Zj6UbSUEzs25+JXtKmmdxni8srJalxjugldeIofnaZ6CWtyeTeBpdXSvoeJvf2uLxSksm9cU7GSlLjTPSS1DgnY6U5Zblmfjiil6TGOaKX5oij+PnkiF6SGjfyEX2SNwC/DOwGvlhVfzzqa0jqzlG8Oo3ok5xIcjnJ41e1H0ryVJKLSe4GqKonq+oDwHuBHx99yJKkzeg6oj8J3As88GJDkh3AfcCtwDJwPsliVT2R5Hbgg8CnRxuupC4cxWtQpxF9VZ0Dnruq+WbgYlVdqqrngVPAkf75i1X1DuDnRhmsJGnzhqnR7wWeGTheBt6c5Bbg3cArgDNrvTjJMeAYwIEDB4YIQxI4itfaRj4ZW1UPAw93OG8BWADo9Xo16jikeWByVxfDLK98Ftg/cLyv39ZZksNJFlZWVoYIQ5K0nmES/XngYJLrkuwEjgKLm3mDqjpdVcd27do1RBiSpPV0Kt0keRC4BdidZBn4SFXdn+Q4cBbYAZyoqgububj70UvdWKLRMDol+qq6Y432M6wz4drhfd2PXpLGbKpbIFijl6Txm/k7TA1+pH36nttGEZY0Nf4+axzc1EySGufNwaUp6DK56gSsRmWqI3qXV0rS+Fm6kaTGWbqRJsRSjKZl5lfdSNuNCV3bjaUbSWqcNweXGH79uqN4bWfW6KV1rJfA/UKTZoU1ejVvrWS9VqLuOjp3FK9ZYY1ekhpnjV5zyxG55oWJXs0wcUsvzclYzYTN1tklfZeTsZqYLiPuwcTtxl/SaFi60bZi4pZGz0SvkbC0Im1fJnptmaUVaTa4jl6SGueIXhtyVC7NNpdX6v8xsUttcXmlAJO71LKmSjfDbjU7D0zo0vxxMlaSGtfUiF7f5acbSS8y0c8ByzXSfLN0I0mNa3ZEP4+lC0fukl7KWBJ9kp8BbgNeA9xfVX81juvMq3n8IyZp6zon+iQngHcBl6vqjQPth4A/BHYAf1pV91TV54DPJXkd8HuAiX5Ia43WHcVL2shmRvQngXuBB15sSLIDuA+4FVgGzidZrKon+qf8Zv/5qXJnRUnzrPNkbFWdA567qvlm4GJVXaqq54FTwJGs+hjwhar6+9GFK0narGFX3ewFnhk4Xu63fQh4G/CeJB94qRcmOZZkKcnSlStXhgxDkrSWsUzGVtXHgY9vcM4CsADQ6/VqHHFIkoYf0T8L7B843tdv6yTJ4SQLKysrQ4YhSVrLsIn+PHAwyXVJdgJHgcWuL66q01V1bNeuXUOGIUlaS+dEn+RB4BHgxiTLSe6sqheA48BZ4Engoaq6sIn3dEQvSWPWuUZfVXes0X4GOLOVi7sfvSSNn3eY2mb8ApSkUfMOUwPcWkBSi5rd1KyL9UbPJn1JrbB004FJX9Ism+p+9C6vlKTxm+vSzaQ50SppGizdjIGlHknbiatuNsktjyXNGu8ZK0mNs0Y/ItbfJW1X1ujHzD8AkqbN5ZWS1Dhr9JLUOBO9JDXORC9JjZtqovfGI5I0fk7GSlLjLN1IUuP8wpQkTdik98NyRC9JjTPRS1LjTPSS1DiXV0pS41xeKUmNs3QjSY0z0UtS40z0ktQ4E70kNS5VNe0YSHIF+Jctvnw38PURhjML7PN8sM/zYZg+v76q9mx00rZI9MNIslRVvWnHMUn2eT7Y5/kwiT5bupGkxpnoJalxLST6hWkHMAX2eT7Y5/kw9j7PfI1ekrS+Fkb0kqR1zEyiT3IoyVNJLia5+yWef0WSP+8//+Uk104+ytHq0OdfTfJEkq8m+WKS108jzlHaqM8D5/1skkoy8ys0uvQ5yXv7P+sLSf5s0jGOWoff7QNJvpTkK/3f73dOI85RSXIiyeUkj6/xfJJ8vP//46tJ3jTSAKpq2/8H7AD+CfhhYCfwD8BNV53zS8An+o+PAn8+7bgn0OefAr6///iD89Dn/nnXAOeAR4HetOOewM/5IPAV4HX94x+cdtwT6PMC8MH+45uAp6cd95B9/gngTcDjazz/TuALQIC3AF8e5fVnZUR/M3Cxqi5V1fPAKeDIVeccAT7Vf/wZ4K1JMsEYR23DPlfVl6rqm/3DR4F9E45x1Lr8nAE+CnwM+O9JBjcmXfp8F3BfVX0DoKouTzjGUevS5wJe03+8C/jXCcY3clV1DnhunVOOAA/UqkeB1yb5oVFdf1YS/V7gmYHj5X7bS55TVS8AK8APTCS68ejS50F3sjoimGUb9rn/kXZ/VX2eNnT5Od8A3JDkb5M8muTQxKIbjy59/i3gfUmWgTPAhyYT2tRs9t/7pnhz8AYkeR/QA35y2rGMU5KXAX8AvH/KoUzay1kt39zC6qe2c0l+pKr+Y6pRjdcdwMmq+v0kPwZ8Oskbq+o70w5sFs3KiP5ZYP/A8b5+20uek+TlrH7c+/eJRDceXfpMkrcBvwHcXlX/M6HYxmWjPl8DvBF4OMnTrNYyF2d8QrbLz3kZWKyqb1XVPwP/yGrin1Vd+nwn8BBAVT0CvJLVPWFa1enf+1bNSqI/DxxMcl2SnaxOti5edc4i8Av9x+8B/rr6sxwzasM+J/lR4JOsJvlZr9vCBn2uqpWq2l1V11bVtazOS9xeVUvTCXckuvxuf47V0TxJdrNayrk0ySBHrEufvwa8FSDJG1hN9FcmGuVkLQI/31998xZgpar+bVRvPhOlm6p6Iclx4CyrM/YnqupCkt8BlqpqEbif1Y93F1md9Dg6vYiH17HPvwu8GviL/rzz16rq9qkFPaSOfW5Kxz6fBX46yRPAt4Ffr6qZ/bTasc+/BvxJkl9hdWL2/bM8cEvyIKt/rHf35x0+AnwfQFV9gtV5iHcCF4FvAr840uvP8P87SVIHs1K6kSRtkYlekhpnopekxpnoJalxJnpJapyJXpIaZ6KXpMaZ6CWpcf8LLCVK9XnuZ6sAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "diff_frac = cumu_diff_genes[uptri] / cumu_passed_genes[uptri]\n",
    "_ = plt.hist(diff_frac, bins=100)\n",
    "plt.yscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x11b370290>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD8CAYAAACb4nSYAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi40LCBodHRwOi8vbWF0cGxvdGxpYi5vcmcv7US4rQAAIABJREFUeJztnX9sY9l1379HonYVRSlVSBOgWe1GCrgCuokF1R54bKFS0rpR1iN0t0Gceu0acjqLbn50W7ROg8oqoLoqMJHTNkaKGEi3sVqPg8Z2DbQYQEpXaDeFBMtWdxyrStauBcYzsekk8EqxWKiCYlFz+wffeTq8eiQfR+QjKX4/wEDk4yXffdToe88759xzxDkHQgghnUFXsydACCEkOSj6hBDSQVD0CSGkg6DoE0JIB0HRJ4SQDoKiTwghHQRFnxBCOgiKPiGEdBAUfUII6SBSzZ6Az9DQkBsZGWn2NAghpK340pe+tO+cu1ZtXMuJ/sjICO7du9fsaRBCSFshIn8UZxzdO4QQ0kFQ9AkhpIOg6BNCSAdB0SeEkA6Cok8IIR0ERZ8QQjoIij4hhHQQFH1CCOkgKPqEENJBUPQJIaSDoOgTQkgHQdEnhJAOgqJPCCEJMbeyXfKzGVD0CSEkIe7cuhH+bJbwU/QJIaQJ6AKQNBR9QghJiGa6dRSKPiGEJEQz3ToKRZ8QQhKkWW4dhaJPCCENptnWvYWiTwghDabZ1r2Fok8IIR0ERZ8QQjoIij4hhHQQFH1CCOkgKPqEEJIQrZDFQ9EnhJAEmFvZboksHoo+IYQkQCsIPkDRJ4SQhmHdOa3g2gEo+oQQ0hB8d44+nlhab9aUAFD0CSGkIZQT+Z3FmWZMJ4SiTwghDWJuZRs7izOYWFpHZmEtXAAyC2tNm1Ms0ReRZ0XkayKSFZH5iNcfF5HPBK9vi8hIcLxHRD4pIr8vIl8VkQ/Xd/qEENJaWN+9llLeWZxB9vZN5I9PMTq/isnMYNPcPFVFX0S6AXwcwLsBPAPgfSLyjDfsRQDfcc5lAHwMwEeD4z8N4HHn3FsAvA3Az+qCQAghVwkV+91c/sLzkflVjMyvwgHo7hIAzXPzxLH03w4g65z7unPuuwA+DeB5b8zzAD4ZPP4cgHeJiABwAL5XRFIAvgfAdwH837rMnBBCWgQbtFUx39zbx9zKNg6PTyFmbOGhw8befuta+gCeAPBN8zwXHIsc45wrAMgDGERxAfh/AP4EwDcA/Gvn3J/5JxCRl0Tknojce/PNN2u+CEIIaSZROfj3l2exsbcPoGj9+owPpxs8q2gaHch9O4AzAD8AYBTAL4rID/mDnHOvOOeuO+euX7t2rcFTIoSQ+qPunLmVbYwG7pxKbAR3AkkTR/S/BeBJ83w4OBY5JnDlpAEcAHg/gP/mnDt1zn0bwOcBXL/spAkhpNVQa//OrRtI9/Vgemwo1nuSFv44ov86gKdFZFREHgPwAoC73pi7AD4YPH4PgNeccw5Fl85fBwAR+V4A7wDwf+oxcUIIaUUmltYxPpzGbi5f4suPohn1eKqKfuCjfxnAqwC+CuCzzrk3RGRJRJ4Lhn0CwKCIZAF8CICmdX4cQL+IvIHi4vEfnHO79b4IQghpBeZWtjE+nA6FPMqXrzxYnr0g+ElY/VI0yFuH69evu3v37jV7GoQQEgvfWs8srGEyMxgGccuR6hJkb9+s2zxE5EvOuaruc+7IJYSQS6B+eU3BzN6+id1cHgN9PRXfd/awOQZ3qilnJYSQK4S19OdWtnF0Uqhq7aerLAqNgpY+IYRcApuqObeyHe7I3coeVHzf0Umh4XOLgqJPCCGXwKZqAsVNV9nbN9Hfm0Kqq3z+TrPcOxR9Qgi5BH6Btd1cHpmFNRwen6JQQdgdmtNYhaJPCCGXwPfn7yzOoPDQVbTy7XuTrsFD0SeEkDpgxbu63BdTNnWRSBKKPiGEPAJ+/9udxRlsBtUzHVDRtQMA/b0pbAbZPRoETgKmbBJCSB0YmV/FQF9P7Kycw+NTDPT1JF6KgZY+IYQ8ArZYmqZpjg+nUXjoYrl3FH1vUlD0CSGkBvxsHQDIH58i1SWhgMdJxhQUG64kXVefok8IITXgu2IyC2vh7tpaBFxTNluuyiYhhJDyTGYGcXh8CgBVi6z5JO3aASj6hBDyyGia5kBfDwoPXdUiaz7NaJlI0SeEkEdEc+yPTgqYHhsKLf44DPT1JO7aASj6hBBSFzZrdO0cHp8mvhsXoOgTQsiluHPrBvp7U3BAze6dw+PTluyRSwghJMCK9MTSetgiMdUlyNfg3gGa4+Kh6BNCSA3opqyoujm1FktuRk19ij4hhNSAn1u/lT3AVvbgkerj9/eWVsJJwtXD2juEEFIDtmmKinS14mq1fnYjoaVPCCFVKGeBb2UPLljrtcA8fUIIaUGiLPC5lW1MZgbD4G2cpimWWsfXC4o+IYTExK97f+fWDUyNDYU7cmthMjPYlM1Z9OkTQkgMbAB3dH4V95dnw81VRycFCGrL3tnN5TGxtM7OWYQQ0urYpuZaQ7/WUO7h8Wnigg/Q0ieEkJp5sDwLoFhWudbKms2Glj4hhMTA+t/VraOZO48SlK21ZEO9oKVPCCExsS6dzMIasrdvYmR+9ZHy9Jvh2gFo6RNCSCw0kHvn1o2S5ifTY0M19cTV9yRdaE2hpU8IITGw7h3dVDWxtI6jkwK6u6Qma78ZqZoKLX1CCKkBa6GPD6fR35uqSfCbaeUDFH1CCKmJ3Vw+dPHs5vJhjn4cNODb8pa+iDwrIl8TkayIzEe8/riIfCZ4fVtERsxr4yLyBRF5Q0R+X0R66zd9QghJlvHhdFhDf2dxpqYc/Wp3BEncAVQVfRHpBvBxAO8G8AyA94nIM96wFwF8xzmXAfAxAB8N3psC8FsAfs4598MAfgxAbV0GCCGkifhCrJb+Vvbgkdod2uqcUa81GnGu8sojIu8E8BHn3E8Ezz8MAM65XzZjXg3GfCEQ+j8FcA3FheL9zrkPxJ3Q9evX3b1792q+EEIISRoV/VoaogPnm7vqiYh8yTl3vdq4ONk7TwD4pnmeA+AvR+EY51xBRPIABgGMAXDBonANwKedc78SMdmXALwEAE899VSMKRFCSPJMLK2HmTt3bt1A/vi05vILADAyv4pUlyB7+2Z9JxiDRgdyUwD+KoC/E/z8SRF5lz/IOfeKc+66c+76tWvXGjwlQgh5NHRD1ebePjILa48k+Mpl6vBfhjii/y0AT5rnw8GxyDGBeycN4ADFu4IN59y+c+4YwBqAt1520oQQkiTqg59b2cZuLo+psSH096YuVRNfXUJJp2/GEf3XATwtIqMi8hiAFwDc9cbcBfDB4PF7ALzmisGCVwG8RUT6gsXgRwF8pT5TJ4SQZNDg651bN3B0UsBW9gCHx6d1aZOYdPpmVdF3zhUAvIyigH8VwGedc2+IyJKIPBcM+wSAQRHJAvgQgPngvd8B8KsoLhw7AH7PObda/8sghJDGU2+rfHQ+eTmM5VRyzq2h6JqxxxbN4xMAP13mvb+FYtomIYS0JerWOTop1K0JOlBb05V6wR25hBASwcTSemjZ7+byyB+f4qyOgt8sKPqEEBKBZurMrWxjZ3EGU2NDTZ5RfaDoE0JIFeZWtrGxt98Qd4yNE7REGQZCCOlUtG7+5t7+pdIzK2Gzd5LI5KHoE0KIh7W4d3N5OFQvltYuUPQJISTABm6t8Deyn23Sm7PYOYsQQgJ0E5YGcTMLxUz1q2LlA7T0CSEkErXAGy34LbcjlxBCrjJR9fKBohg3Oi+/UcHhSlD0CSEdjW1qoq6dzMJa6NppJM1wG1H0CSEkQF0tk5lBnNXQBrGdoOgTQjoeFXvthLWVPUB3E1wvSUDRJ4SQAG163t+bwmRmsNnTaQgUfUIICdCm54fHp9jKHjR7Og2Bok8IIQGanz89NnSlcvMtFH1CCPHY3NtHEh79ZkQNuCOXEEICNJCb7usJe9g2kvvLsw0/hw8tfUJIx6N5+uPDaRwenyYi+M2Cok8I6XhsKYTphJulJF1wjaJPCOlo5la2S3bfJuXPV5KuvUOfPiGko7Ebs/LHp4n58wf6ejC3sk3RJ4SQpJlYWsfRSQEOQD4hf76mhyYNRZ8Q0rHMrWxjN5dH/vgUU2ND2GxQH9xWgj59QkjHYKtpWteKAxrW+LwcE0vriQdxAVr6hJAOQUXeF9pmpWc2y71DS58QcqWxVr3+3M3lsZU9wMj8alN2xeq8mgFFnxByJVFRVReOPh+dX0X++BRnDx0G+nqa5sPXRYiN0Qkh5JKoRT+xtI7x4TS2sgfI3r6JuZXtsE7+2UPX1J23zUjXBGjpE0KuEL51Pz6cviCsZw8dCi3QFatZpZsp+oSQK4Pf71abnOvzpDN0KpG9fbMp56XoE0LaHusXV+Hf3NvH0UkBI/OrOHvosLG338QZtg4UfUJI26MWvXXvOBRdOQBaxrq3aBnnpKHoE0LanqOTQsnzkflVDPT1oLtLIGhOs5Jq7CzOXMjcSSKTJ5boi8izIvI1EcmKyHzE64+LyGeC17dFZMR7/SkRORKRf1KfaRNCyDmTmcHQhz+xtI7psSHkj0/DloetaOlHZe8kkc1TVfRFpBvAxwG8G8AzAN4nIs94w14E8B3nXAbAxwB81Hv9VwH8zuWnSwjpVDSn3QZqtSTyxt4+trIHODw+RT5oaq6pma0o+EDRJdWMDVpxLP23A8g6577unPsugE8DeN4b8zyATwaPPwfgXSIiACAifwvAfQBv1GfKhJB2plahs356u9FqN5cPM2AG+noAAKkugQNQCNIyW5mdxZmWzdN/AsA3zfNccCxyjHOuACAPYFBE+gH8UwD/otIJROQlEbknIvfefPPNuHMnhLQhtQpdOYv46KSAuZXtMCDaDkIfxVXrnPURAB9zzh1VGuSce8U5d905d/3atWsNnhIhpB3QKpS2MJkvkFoWuZ172iZt7ccR/W8BeNI8Hw6ORY4RkRSANIADADcA/IqIPADwjwAsiMjLl5wzIaSNiGvJ+uP8KpQTS+vYzeWxm8tjY28fhaCMgtr2rZihU4lWLq38OoCnRWQURXF/AcD7vTF3AXwQwBcAvAfAa845B2BKB4jIRwAcOed+vQ7zJoS0CXEtWd9fnz8+xf3lWQDFkgWTmcGKpQvazbETVSIiCapa+oGP/mUArwL4KoDPOufeEJElEXkuGPYJFH34WQAfAnAhrZMQ0tlYq1Yfq7WrfnnNxtlZnMHU2BDmVrYxOr8K4HxRmMwMJjnthtGsHcJSNMhbh+vXr7t79+41exqEkAZgc9O1AiYAbO7tY2psCEDRT7+zOBO+vpvL4/D4FKkuactAbTkEwNTYUEmd/0t9nsiXnHPXq43jjlxCSF0o55+OqoszsbSO/PEpdnN53Ll1A91dgo29fWzu7WN8OI3R+VXsLM6Egg/gSgk+AKSDNFMg2WAuRZ8QUhfKCVdUs5Dx4TTSfT04PD7FSOC+AYqW71b2APeXZzEyv9rWWTnVUJ8+m6gQQlqeau4IrXKpLhuLFkc7PD6FAKH4A+c15ptVjCxJ9PtrxZRNQggpoZxQaWmE3Vwe95dnsbG3H7pwdnP5MFCbD3z0QHGTlT7u702FqZhXmVSXsEcuIaS98OvgKP29KRydFDCxtB7mzs+tbJcIuZZK8EsmXHWxV/p7m+dkoXuHEPJI2CwboOiSOTophNb60UkB3V2Co5MCdnN5DPT1YHw43fHNTFJdEm48a0afXFr6hJAL+Ba8teqjyiNo6uVkZrAkvVL/afXLThd8xe/lmyQUfUI6nCgXjeIHG+dWtjE+nMZW9gCZhTVkFtZwdFLAVvYA48NpbAairu4a9dUD7bdjtlFMZgYv9PJNEm7OIoSEqLvBt0Q1AOv7ovPHp5gaGwrFvrXUpPVIdUko+vWGm7MIIRWJct34YpRZWMPcyjbOHjqcBX76Q1PV0uG8nAAFvzqtUEKCok9IB+FvkgLOm5OoNW8XABWpbtOcRPGzcUg8muHHtzB7h5AOwM8SmVhaL8m8AUqt0N1cPmwurpuoyOUY6OtpuuADtPQJ6Qju3LqBiaX1UHR2FmdKMnA0rXJ0fhWbe/s4OikAQIkbh1wO7fTVbCj6hLQRlUSjWsEzFfioei/9vSnkjcBbNw6t/PrQCv58gNk7hFw51HUTVd1S2coeoL83hcPj09CFYxEkZ90nea5mMdDXc6ETWL2Jm71Dnz4hbYwNulqLXoVfUd/9oal5o0KvP229+iRF2J5LFwBBMXh8Fcop6w7cZuy+jYLuHULaFO0Zq2gWDoALlv74cBr5ICBra91YWkFgu7sEA309F1xMV4FWEHyA7h1C2oYoS9E/ZoV+N5cPO08dnRRQeOiQ6hKcBYXOWgXr3tH4gVr57e76SXUJsrdvJnIubs4i5IqhLhybT2/dOpmFtXBn7EbQgWore4DD49NwJ22hxQQfOBf16bEhpAMr/ywQ/Kh6/O1EqwRvLRR9QtoEFffs7ZsYnV8NN1RpOmZ/b6pkh+zG3n7FksXNzsqZHhvCdCDqqS7Bbi5fkkEENK95eD1IdUlTOmNVg6JPSJtg3TjdQTA2e/tmWNJYG5MM9PXEEvSkLX5bfE3Ffit7AAHCTCI7ttXuSGpF765axZevUPQJaVH82jjWnaOCYvvIavDz8Pg0FMwB03w7aey5VfAH+nrC45vBnUh3l4S7fnUxa/cgrgAlNfNbCQZyCWkxomrjTCythzXrt7IHZYOxzQ58atqn/1Mtew0qT2YGsbG3H7lH4CrwYHk28XMykEtIG6KlErSv7Fb2ACPzq+Hr2ji8nLBHHbdulXqQ6hKkugQChK4k/akCr5k3/b0ppLok9M1rFpFtjn7VEFTuUdBsKPqEJEiUCNjsG+Xw+BSZhbXQzaG1cfwc+1rk/DLSr8I+0NcTpiBqK8R0Xw/yx6dI9/WEFr19Xee7WSWwfBUY6OvB/eXZC81nWgmKPiEJEpXNsZU9wMTSOs4eutCFAyB0jwDF1L98UDJB0U1MyvTY0AUffrVdtnYhKOf/f7A8i+5gV2n++BSj86thTGEyM4ijkwKmxoawsziDjb19bOztR24Cay1Hcv0Z6OsJC9W1MvTpE5Iwfg0czcDx0xWBUtEGon325fz4fimDVMSGJxUqW3Btyvjf1YJXP/zm3n4YbAXaP+BaLwRAOlg0G11jp+wc6NMnpLlEdabyd8xaHIoiba1vK6q+ZQ9EC36qjCifRQj04fFpScbP/eVZ7Oby2NzbRz44pmN2c/lQ8M+MFd/sfP9mk+oSpIOCas0S/Fqg6BPSIKJcObu5fBiMPTop4OyhQ2ZhLfRxV9ox67sO/EUgZQQZKC4g096OVrsgCC7mzutcHIqWq309H4i/P8dOtvW1zIJ1y7U6FH1CGsDE0joyC2vYyh6UZOGMD6dxZrJXNLd+wBPYSmL9YHkWAoTuGMWv8WKtfLVG9dhAXw+mxoZwFsQNBMVA62RmsKQKp7qDdK6kFP3ONWDbitk6PvTpExLwKKVv9T02j343lw/LG6t13t+bCjNZpseGsJU9wGRmMLT6fYHWhcCWQlasW8WZ9+h59K7BbnZSzh463F+exej8aviavsePH5Bo9Bu934Rc/ErQp09IjUQJfhzLzQo+UCxjPDq/GgZI9dhkZhADfT2hRa3nKxhr2wrv0UkBgmKGTH9vKvwJXMy0UR+7NkXR4mWFhw79vSn096aQvX0T3V0S1tYvPHQ4C96jz0lldCFtNcGvBYo+IR5zK9uYWFovsfxV/NVtY7E7ZYHi4pHu60F/bwo7izPI3r4ZunjGh9PoDoqL6eeo5ajuFj1WeOgwNTYUVs5UpseGQr+71trpNhumdD725/hwGiNBquXh8Wlo5VPm46Hf89TYUGKlkhtFLPeOiDwL4NcAdAP4Tefcsvf64wDuAHgbgAMA73XOPRCRHwewDOAxAN8F8EvOudcqnYvuHdJMooTeD8jaUgKbe/th5oZa/H6qo3XpaGqkbmiydeTTpiSBPs8fn4ZW5cj8Kgb6esKSybbYmt4B2C5YwHlQV7tRUeRrQ1B047RK16tK1M29IyLdAD4O4N0AngHwPhF5xhv2IoDvOOcyAD4G4KPB8X0Af9M59xYAHwTwqfiXQEiy+PXpAVzYWXnn1g3sLM7g7KELBR84b0eo1r7uRgWKwrGxtx8KM4BwJ6sGcHW8um00VdIFnz06v4rpsSHkj0+xubePs4cOcyvbYa18teYVm2WjQt/ugp9Uauj02BAExYB5usyGtXYmjnvn7QCyzrmvO+e+C+DTAJ73xjwP4JPB488BeJeIiHPuy865Pw6OvwHge4K7AkKagu+jV7G2aGBWM2wyC2vhc33cHQh1/vi0JJXS3vpnb99Ef2+qpDzBYWDl6+PD41NMZgZReOgwmRkMffLAedmDo5NCWCc/Hbhy7i/PYit7EI7dzeXDAmZXlVoWLb3TsXWBAJTU71d3mM9uLh9uUNO8+1a38mshjug/AeCb5nkuOBY5xjlXAJAH4LeM+SkAv+ec+3P/BCLykojcE5F7b775Zty5EwKgerA1SuhtE/EoUVcyC2s4e+hwdFLA+HAa2ds3kT8+xVkg0g4osbS1IflkZhCjQdnjo5MCNvf2cXh8GvrjNairqZIDfT3YzeWRChaSB8uz2FmcwfhwGv29qVCc1JWjZRt04fCbnLcL5YrB2UJulcYKzu8A1O9u012nxobCRVPvuIBicHwyMxguyKkuwYPl2bCJ+VUSeZ9EArki8sMounx+Nup159wrzrnrzrnr165dS2JK5ApR7g/Ur0GvY3cWZ0KBt7XpgaIYjA+nQ8HVeu+aXjm3sh1anLu5PATnm6Y0UDs6v4qt7AEcipalpknauW5lD3B0UkB3l4RWpWb4dHcJRuZXMTK/WtI5SoO7KvLt7q4Bztsi+uj12RiHrUWkwq6byB4sz5a0JtT9BhuBK2ygryfcbPYguEvS1FplYmm97YO0cUhVH4JvAXjSPB8OjkWNyYlICkAaxYAuRGQYwH8BMOec+8NLz5iQALWqgdIArB63z49OCphb2S4JwgIIM2Nse77psSGMzK+Gt//9vSns5vJh4NQW1tL3qX9/N5cvcd9o6YK5lW1Mjw1hbmU7bE6uGTQq7BoP0GCvfnarW+9xA8RR+wDi9ATQ70sXh4G+nvC70sA2cN4IXt1zUefQBV7vztohQFtv4lj6rwN4WkRGReQxAC8AuOuNuYtioBYA3gPgNeecE5EBAKsA5p1zn6/XpAmZW9kuW+dERWBuZRujwS7YycxgKPhAUaT92vQPlmfxYHk29I0XjDAfnRSQWVjD+HA6XDS0PaGmVu4szoRuoPHhNKbHhsIduJt7++E5C4Hlr7n3QFHY1LffilZ8pZr8Dhf3DehzDYpqjX2NU0TFHvQcafNePaa/B70TU5eX3rXZzCkEx9Xat66yncWZ8P/OnVs3WrKHbaOpKvqBj/5lAK8C+CqAzzrn3hCRJRF5Lhj2CQCDIpIF8CEA88HxlwFkACyKyE7w7/vrfhXkSlOpIYV14fhjbUBO0yjVl9ttxER9uSPzq5hYWi8ROE15VIHezeVDa7G7S9Dfm8KDoEjZ3Mp2yaYrZWpsCFNjQ2HJBT0XcG7lFx66lmwCrt9FVHE1+9i/G8kH7Q839vbhUFxk1RV2dFLAzuJMZFlnjWnoJjbrepsaG8JuLh8u3lNjQxidXw0Lnang2/8T9v36WlRKbifBMgykJYly1/h589ZyV4v5zq0byCyshWUPgPPaNtZlkgoEG0CY967vsa38bA68fl729s3w/BabeWOtTgAl525lKtXZ0e9MvxMtGKf7C9QtY/cY6E9b8sH/LnS8fq7vshsJ0lV9941+3/pd+z1pO07MWYaBtAu+BW+f6227Htvc2w/9teq26e9NYXNvPwzO2tIC/b2pkmwOFWX16QIIrXd1G9y5dQMpk2efPz4N3QWalaOCYtM1p01AdjeXL8msaXbf2nL4lruWdwYQ7iHQrBht8qKuLXVr6TnsfgAVYN2PMJkZDKtR2oV3wMQ/1HXj//6136zdh6AZTnpMjYIoS56UQkufNI2ozBpFxV4Lk6l15+8+9RuCWKtSsVkd+nkAQmG2u2XVklRLXq1JPRdQtErHh9Nh8FYt11Zo8u03TgGiA6i+20ZRi1qv199VDKDE0rbFx/R3pp2zuoMArLXitbCb/a6VuZVtbO7thwuJ/h70fLrrGUB4R9Zp1nwl4lr6FH3SFFRU7R+yX5Nc3Tdnwc7SASNAD0ylSHXfAOeWt+0EpdvobQ0bFSLgXPhs9Ut1zehn2EBis3zvcUoc24UnavyD5VlkFtZKrkODyBt7+2GmzNTYUPi6FXAAkQvdgHHn+C63KDecXXzLLfyKFX7r0okaazO6Og26d0jTsUXK9B+A0D1ib+WPTgrYCFw0QDEAmA9KDGgetm6KAoqbpqbGhkLhUBcMgHAzlQr1aBA01WCqipa6hjSbZDeXD/Px88Yto6KZdLDVd8qkg1xz26Rc/z1Yng3FUFNNbQmBVNBQZXR+NbwOtdy1htD02FBJ4FM/eyrIvNE5aD0gAcI7JS0kB5wHvzUXXje1aeqr1g3S7JmJpfVwrGKL3lmX3nSQJaWZVz429kOioaVP6o7vV42qZ5NZWAsbdKi7xAqqdUlYt43mt/sBQRt0tValio4Kio7R81nLVvu/NrPEsF63zeXX78nebdiibn4dfxVKdbnYfQn6PqDU/aLo92zFX91sir5mi8yV86P7x20wVo/7ln6l3PlOtuSrQUufJI7tBavPgeIfqpYW1tRGv7SApvQB57sv1aK1gVAtWewXLNPAot9S8PD4FBt7+xgfToduj3ywaQooCpjWtbEplUmg51HLXK9bLW4rxjb7Ra/BBkyBUitX0xltDEQFvzvooqVlCPLHp+H+A7svQY9poFutdT2ueyB0p7KK9W4uH1rxQKkRoPnxOs+t7EH4/6OS4NuyGVEpvLTu40PRJ5eiXMPvcoXN1MrUwmMqtNqU26KCpBKsWTk2KKgt/awVrGN3c/mSgmTWqq9Uq+aylr6/8UjMP3Wz2IVFxXcq2MgEIMxlV/eTDUYDRVFPBTu5+eIfAAAQuklEQVR9NXcdOLesdSfwZGawZCHUTWG6W3hncQbpYIdr9vbNsKaQul8mM4PYWZzB4fEp5la2S4LrU2NDuHPrxoXSBX4Gji/m9v9M3LIHdiGIWhQY0I0PRZ88En56HHBeRkBFxjYOsWULtrIHYeEx9RsLEG5YcigGHPU9tuGH73fWlEK9c9BaK7pg6CKjAWC16uuJirwKtu7U1X+ajZIO0h6BYq6/BqR1ftp8xW4yUkve7iPQ+Z95i5P6wdWFoguFbcuo1rn681XIJzODmFhaLykVbX+/04HAazkKAJGWvB7X59ooxhflzb39cDGw+fX22qOwd5MU+keDok8iicqd1z9g/ePcCAJsKtaaQqc+5xETNNR2fgOBe0Ytb5vHbkvgar69+p31uGaC2KAtUGqx21r0NgjbKHRh0U5WDgitZrWYtUyDC65hNKjto6iQqn/e96nb8gP2TkIzktRtoiUgrMVvLXa1zq1g6p2C7mz1ewjoY10g7HOLL8ITS+vhYgKUCrZtTOIXw6sGxf5yMJBLYhFV3EyDsTbn2y+WZdMG/RRCG6yNKrIFnGeWqKsCOL+j0FROn0Y1+LZzss3Q1QqPKv6mi5fNYwdKA5qZhbVwl69vPdvm6lrwze4liAqgqqsmylq2gmnPq/OJGmfFuZx1D5T2JtDgsV/J0o5V/3y14K0/DxIN8/RJbKL+mFQwNLXPFwXrbrBZOP4WeyvAlcT4wfJseMdgH1sa2e5PP1vnqM81c0hdNIq/YUgzhI5OChUF1y/f4Aui/V2Mzq9WbdWnOff6+eU+ywq7XWh88Y1ibmU7XEgqfX7UHOw8gfMFM2o8Rf1yMHuHxCYqwKbHBoyPWfPobSEs61e2WTa2R6tm2MB7zZbKHZlfDQOdKvh+Dk09BN//TM2ciRJ8jSuon3sre1DyXdmmKXdu3QhLCfj4mSnqRrGLqN/Ba2Jp/cLO1KiFpFIg1A+o2wVmMjOIuZXtkiBvOT+6uof8ceX8+Dp//aeF6NTVFTXe/zzSOOLU0ydXjKjceZuHrT5nW0hLhU9FUC3bqIbbgvPmGLbZt7/1368X74v6o4q8L+KKL+p6BwMg3AimKY3qi9aNXkBRKNVCtmiBNw12Kr4Va4OUmkevz63bB8AFF45+Vjn3TlTJAivy9i7Dt9DL5b77rpdqdwP2Nev7J60F3Tsdgv9HqW4B+8dva6ZYl4ytqGip9D8njl/dVmWs9/9CTc+8b8oOABdr7wDnLRNVzHW7v2bWlNuAZDdD2ec+fjqrHwytVHrAt6r9TUz+a5VcPHotcYKlWuLCxirKnYO0BnTvkBLUAlSh0Awa4Nxnq1khKoq2PZ/izL9y6Pv0cTmc97MW7EYu3cSlpQimg/IMU0GnKuvz1vRFFd/x4XToTlIXys7izAXBt354u9HIWuJ+fro/Hji34KMyZHQO9njUJicr+H5553LnnlvZxkhQe96ex0dLVqjg22sqFzi2z/XOgbQuFP0rRtQfnRV7RXeoah9WFfd0X0+YB64uEnXX2JIHUUQV+LJZO4+Kfb+W41XXiG5qAop3L1qXRXPK9bF1X9lF4M6tG2Huf5SfWUXSduPS99qf9vOiKCfy9jPLvR71ml1o9DN0Yfc/V3/aEtN2rB03NTYU3vXotUfdkUT5423MgrQudO9cQaxrwKbvaas+vxyx7nC19W3uB1Usy/3vUPdJI1MmdW5+WqEVLD/NUP3rfkaKtYztZ/iNsf2CXXFdGeX69VaiWvZLpfnb+el3U+v5Fc2sUcvelrOmO6d9YMrmFadSnrQtGazWnea624bculmqXA14X7yjarXX4o/3ffg24OqnD1pxjpMn7vvNq6Uh2vf5n1ku/TAqPdIX3Gqf7VMpiOpTKR0yTjkDO65aULbS+UhrQtG/4vgC5QcIrfWm5Q40px4ozaSJEn5b7bHcBio9BkRvuALOrXW9K9Ca9TZoWinTo9zi5gtlLdUXa7VeKwVIy90V6J2V3+ijnF+83MIR9Z5Kc49aDGu5W2HWTftC0b9CRFleulPT1pmxDSzsxilf1G0Ovd2A5Lt8fBdNJavebmbysZkvfiaMXp9vYVezuv33thq1bj6qZvHXcsdCOhOK/hXBz+nWY8D5rlgAoejb9n+FiFx561oBoksmdFex7oFzn76OnQ46LWkjDwCxhN6/pnJCHydXPOq7q8UqToLLWOKVPqfW18nVgymbbYifbTFqqk6Ozq+GOxz1j1mF2e8L67cF1KqPwHnFSq2pbtFUTN1YpQuGPp4OSv9Ojw0he/tmmNanm5y0g5Jmzii2iJdSTvD916MWgzjE9Vn733nUuErHNc3TPrb/fOolxJfJAiKdDUW/hdA/VBXMdF9PmFI4FQgqcHHLvrazU2veWuqHXi0ctf51vG3kobjgM+4vz4Zt8O4vz2I3lw/nNLdSrIWuIm/T+4DzrkqWciJaaUGIylEv93n2fJUWiajP911ndnH1r0E/236+bQSin1cup91+VqVSwnEXnajrIqQcdO80Eeu68f94VVxG5ldLCp7pT78mfFTJAfXZx8mw0Z232v/UZtDY0gB2fvpYr8V/HpUVE5UqGTXmssSJBfjBU1007IJlU0arLQJKremahNQD+vRbmDgZKzbtUCtdbgYt/aIYMP58Ha8bkjSoq/5/DQD7Nc2j0iTj4I8tF5SsVuMlLpcRzFp84dXy5KMyeiz+d+LHNQipJxT9FqCSdQicC4ha+1rkzM/AUQs8CnXl2KwZ2/Rb89+B0v6iUaVy486/FYmT5eIvcMDFzCK/TEHU5qdqn09IM2AgN2HKBe1GTV0X7TKlnYpU3KfGhjCxtI77pkVg4aELLUobqHXBeKC0/Z42stYdlQ+WZ0vquvsVF+0mnXLX0yoCFsff7fv7dZ+ClqCwQq8xCLuzFcCFHa02xlJJ8LmZibQTFP1HxM/6sNaiHp9b2Q4FfWNvH+PD6TCl0fYw3c3lkT8+xej8auiCUcu9vzeFqSBb5v7ybFi+V+/P0kGBMX3fnVs3LtR09wXfUk6oqmW+xPluKo2PmzUT5TP3s2WAi8Ft+x1Ebe7S41HX6R+rZuFT7Ek7QffOJfFFyQY/1TVjA7AW62KwKY62u5GK1+befkn3pqggqf+avl7ueJRPup4CVmtwttL4WvLbaXmTToQ+/TpSTlijRMr6zNXXroHTqSC4aq1Qf6u+v3hEWamVWuxVmnu9Rb5c0NPOs1o7vloWpXrMu5XcVoTUE4p+jURVSbQCodkXGnT13Qk24KqbnnS8bXsXdb5Kc6qW8VFNzCuJnJ1DnPlUE+5K2Sy22mc5H3lUQTCKNCHxoOjHpFxJXCtO6mbxSxvYAKtflhYoL/Jxzum/Vok4Lh5/8Son0Frawbbvi0pRtGmnis0IolgTkiwU/QpEuSKAc9+6irkVP60roy4L21Bax5XbjKTniKrx7m98AqJ7q8atfliuAqVuMrJi7TcW0Wu3dX78edsNVsw7J6R1iCv6sRqji8izAH4NQDeA33TOLXuvPw7gDoC3ATgA8F7n3IPgtQ8DeBHAGYB/6Jx7tYbrqDu+aAHF7IyR+dWw5rz63lXYNeddc+W1Wbg20AZKBdQGYX3r3d+RurM4E6Z1Kn6DELsg+C4TewcCnHdHsouL3Zhlg8ZxarD75RX0J7sjEdKeVLX0RaQbwB6AHweQA/A6gPc5575ixvwCgHHn3M+JyAsAftI5914ReQbAbwN4O4AfAPDfAYw5587Kna+elr76iP3mHABKSgmH1xH8tBue1GUDICxjbDdP6UYoLU885e2ItSWO/V2ytpyCrX0TJcZRFj9dKIQQpW7uHRF5J4CPOOd+Inj+YQBwzv2yGfNqMOYLIpIC8KcArgGYt2PtuHLnq4foZxbW6tKurxy2AYm1+NXifmDKGwBsTkEIaTz1dO88AeCb5nkOgK9c4RjnXEFE8gAGg+Nf9N77RMRkXwLwEgA89dRTMaZUnhHPVVILuvGpHnne9r10hRBCWoVYPv1G45x7BcArQNHSv8xnPTBByEeF1jgh5KoSpwzDtwA8aZ4PB8cixwTunTSKAd047yWEEJIQcUT/dQBPi8ioiDwG4AUAd70xdwF8MHj8HgCvuWKw4C6AF0TkcREZBfA0gP9Vn6kTQgiplaruncBH/zKAV1FM2Vxxzr0hIksA7jnn7gL4BIBPiUgWwJ+huDAgGPdZAF8BUADw9ytl7hBCCGksHbk5ixBCrhqsp08IIeQCFH1CCOkgKPqEENJBUPQJIaSDoOgTQkgHQdEnhJAOgqJPCCEdBEWfEEI6iJbbnCUibwL4o0t8xBCA/TpNp13otGvutOsFeM2dwmWu+Qedc9eqDWo50b8sInIvzq60q0SnXXOnXS/Aa+4UkrhmuncIIaSDoOgTQkgHcRVF/5VmT6AJdNo1d9r1ArzmTqHh13zlfPqEEELKcxUtfUIIIWVoW9EXkWdF5GsikhWR+YjXHxeRzwSvb4vISPKzrB8xrvdDIvIVEdkVkf8hIj/YjHnWk2rXbMb9lIg4EWn7TI841ywifzv4Xb8hIv8p6TnWmxj/t58Skd8VkS8H/79vNmOe9UJEVkTk2yLyB2VeFxH5t8H3sSsib63rBJxzbfcPxQ5efwjghwA8BuB/A3jGG/MLAH4jePwCgM80e94Nvt6/BqAvePzz7Xy9ca85GPd9ADYAfBHA9WbPO4Hf89MAvgzgLwbPv7/Z807gml8B8PPB42cAPGj2vC95zdMA3grgD8q8fhPA7wAQAO8AsF3P87erpf92AFnn3Nedc98F8GkAz3tjngfwyeDx5wC8S0QkwTnWk6rX65z7XefccfD0iyg2oW9n4vyOAeBfAvgogJMkJ9cg4lzz3wPwcefcdwDAOffthOdYb+JcswPwF4LHaQB/nOD86o5zbgPFtrLleB7AHVfkiwAGROQv1ev87Sr6TwD4pnmeC45FjnHOFQDkAQwmMrv6E+d6LS+iaCm0M1WvObjtfdI5t5rkxBpInN/zGIAxEfm8iHxRRJ5NbHaNIc41fwTAB0QkB2ANwD9IZmpNo9a/95qo2hidtBci8gEA1wH8aLPn0khEpAvArwL4mSZPJWlSKLp4fgzFu7kNEXmLc+6wqbNqLO8D8B+dc/9GRN4J4FMi8iPOuYfNnlg70q6W/rcAPGmeDwfHIseISArF28KDRGZXf+JcL0TkbwD4ZwCec879eUJzaxTVrvn7APwIgP8pIg9Q9H3ebfNgbpzfcw7AXefcqXPuPoA9FBeBdiXONb8I4LMA4Jz7AoBeFGvUXFVi/b0/Ku0q+q8DeFpERkXkMRQDtXe9MXcBfDB4/B4Ar7kgStKGVL1eEfkrAP4dioLf7n5eoMo1O+fyzrkh59yIc24ExTjGc865e82Zbl2I8//6v6Jo5UNEhlB093w9yUnWmTjX/A0A7wIAEfnLKIr+m4nOMlnuApgLsnjeASDvnPuTen14W7p3nHMFEXkZwKsoRv9XnHNviMgSgHvOubsAPoHibWAWxaDJC82b8eWIeb3/CkA/gP8cxKu/4Zx7rmmTviQxr/lKEfOaXwUwIyJfAXAG4Jecc+16Bxv3mn8RwL8XkX+MYlD3Z9rYgIOI/DaKC/dQEKf45wB6AMA59xsoxi1uAsgCOAbwd+t6/jb+7gghhNRIu7p3CCGEPAIUfUII6SAo+oQQ0kFQ9AkhpIOg6BNCSAdB0SeEkA6Cok8IIR0ERZ8QQjqI/w+eOuHInN/MkwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.scatter(diff_frac, divs, s=0.6, linewidth=0, rasterized=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('/Volumes/Botein/uhgg/pairwise_distances/{}/pair_snvs'.format(mgnify_accession), cumu_pair_snvs)\n",
    "np.save('/Volumes/Botein/uhgg/pairwise_distances/{}/pair_passed'.format(mgnify_accession), cumu_pair_passed)\n",
    "np.save('/Volumes/Botein/uhgg/pairwise_distances/{}/diff_genes'.format(mgnify_accession), cumu_diff_genes)\n",
    "np.save('/Volumes/Botein/uhgg/pairwise_distances/{}/passed_genes'.format(mgnify_accession), cumu_passed_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "cumu_pair_snvs = np.load('/Volumes/Botein/uhgg/pairwise_distances/{}/pair_snvs.npy'.format(mgnify_accession))\n",
    "cumu_pair_passed = np.load('/Volumes/Botein/uhgg/pairwise_distances/{}/pair_passed.npy'.format(mgnify_accession))\n",
    "cumu_diff_genes = np.load('/Volumes/Botein/uhgg/pairwise_distances/{}/diff_genes.npy'.format(mgnify_accession))\n",
    "cumu_passed_genes = np.load('/Volumes/Botein/uhgg/pairwise_distances/{}/passed_genes.npy'.format(mgnify_accession))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "core_gene_file = '/Users/Device6/Documents/Research/bgoodlab/UHGG/E_rectale_core_genes.faa'\n",
    "# core_gene_file = '/Users/Device6/Documents/Research/bgoodlab/UHGG/B_vulgatus_core_genes.faa'\n",
    "pangenome_file = '/Users/Device6/Documents/Research/bgoodlab/UHGG/E_rectale_pan-genome.faa'\n",
    "core_gene_name_file = '/Users/Device6/Documents/Research/bgoodlab/UHGG/E_rectale_core_genes.txt'\n",
    "\n",
    "fasta_sequences = SeqIO.parse(open(core_gene_file),'fasta')\n",
    "# fasta_sequences = SeqIO.parse(open(pangenome_file),'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'UHGG_utils' from '../UHGG_utils.py'>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reload(UHGG_utils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished 0 at 12:26:55\n",
      "Finished 100 at 12:29:07\n",
      "Finished 200 at 12:31:09\n",
      "Finished 300 at 12:33:01\n",
      "Finished 400 at 12:35:24\n",
      "Finished 500 at 12:37:52\n",
      "Finished 600 at 12:40:16\n",
      "Finished 700 at 12:42:34\n",
      "Finished 800 at 12:44:53\n",
      "Finished 900 at 12:47:29\n",
      "Finished 1000 at 12:49:47\n",
      "Finished 1100 at 12:51:20\n",
      "Finished 1200 at 12:53:27\n",
      "Finished 1300 at 12:56:00\n",
      "Finished 1400 at 12:58:21\n",
      "Finished 1500 at 13:00:28\n",
      "Finished 1600 at 13:02:50\n",
      "Finished 1700 at 13:04:34\n",
      "Finished 1800 at 13:06:19\n",
      "Finished 1900 at 13:08:24\n",
      "Finished 2000 at 13:10:49\n",
      "Finished 2100 at 13:12:52\n",
      "Finished 2200 at 13:15:16\n",
      "Finished 2300 at 13:17:09\n",
      "Finished 2400 at 13:18:59\n",
      "Finished 2500 at 13:20:44\n",
      "Finished 2600 at 13:22:40\n",
      "Finished 2700 at 13:24:20\n",
      "Finished 2800 at 13:26:36\n",
      "Finished 2900 at 13:29:05\n",
      "Finished 3000 at 13:31:03\n",
      "Finished 3100 at 13:33:19\n",
      "Finished 3200 at 13:35:08\n",
      "Finished 3300 at 13:37:11\n",
      "Finished 3400 at 13:39:34\n"
     ]
    }
   ],
   "source": [
    "coverage_array, gene_files = UHGG_utils.compute_gene_SNV_coverage(grouped_snvs_base, genome_mask, debug=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "species_core_genes = UHGG_utils.get_species_core_genes(coverage_array, gene_files)\n",
    "np.save('/Volumes/Botein/uhgg/core_genes/{}/SNV_coverage'.format(mgnify_accession), coverage_array)\n",
    "json.dump(list(gene_files), open('/Volumes/Botein/uhgg/core_genes/{}/gene_files.json'.format(mgnify_accession), 'w'))\n",
    "json.dump(list(species_core_genes), open('/Volumes/Botein/uhgg/core_genes/{}/core_genes.json'.format(mgnify_accession), 'w'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([ 886.,  178.,  130.,  102.,   97.,  104.,  176.,  163.,  311.,\n",
       "        1282.]),\n",
       " array([4.10593307e-04, 9.79675631e-02, 1.95524533e-01, 2.93081503e-01,\n",
       "        3.90638473e-01, 4.88195442e-01, 5.85752412e-01, 6.83309382e-01,\n",
       "        7.80866352e-01, 8.78423322e-01, 9.75980292e-01]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAD8CAYAAAB+UHOxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi40LCBodHRwOi8vbWF0cGxvdGxpYi5vcmcv7US4rQAAEI5JREFUeJzt3XuM5WV9x/H3R1awXkF2Sujutkvj2pbSNpIJxZhY6xoFNCxJ1UC0rnbTTS1aK6a61j9oNCYYW60mVrt1qUtjUUpt2VQsJYAhbbrUQSxy8TJFhN2CjHLphXhBv/3jPNhx3WXOzjlzDrPP+5VM5vk9v+f8ft9nZjKf+V3Ob1JVSJL684RpFyBJmg4DQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktSpNdMu4LGsXbu2Nm7cOO0yJGlVufHGG79ZVTNLjXtcB8DGjRuZm5ubdhmStKok+fow4zwFJEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnXpcvxNYkqZt445PT2W/d1700hXfh0cAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4tGQBJLk5yX5JbFvW9N8mXktyc5O+SHLto3duTzCf5cpKXLOo/o/XNJ9kx/qlIkg7HMEcAHwPOOKDvauCUqvpl4CvA2wGSnAycC/xie82fJTkqyVHAh4AzgZOB89pYSdKULBkAVXU9cP8Bff9UVY+0xb3A+tbeAnyiqr5TVV8D5oHT2sd8Vd1RVd8FPtHGSpKmZBzXAH4L+ExrrwPuXrRuX+s7VL8kaUpGCoAk7wAeAT4+nnIgyfYkc0nmFhYWxrVZSdIBlh0ASV4LvAx4VVVV694PbFg0bH3rO1T/j6mqnVU1W1WzMzMzyy1PkrSEZQVAkjOAtwJnV9XDi1btAc5NckySk4BNwL8BnwM2JTkpydEMLhTvGa10SdIolvx/AEkuBV4ArE2yD7iQwV0/xwBXJwHYW1W/U1W3JrkMuI3BqaHzq+r7bTtvAK4CjgIurqpbV2A+kqQhLRkAVXXeQbp3Pcb4dwPvPkj/lcCVh1WdJGnF+E5gSeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSp5YMgCQXJ7kvyS2L+p6Z5OokX22fj2v9SfLBJPNJbk5y6qLXbG3jv5pk68pMR5I0rGGOAD4GnHFA3w7gmqraBFzTlgHOBDa1j+3Ah2EQGMCFwK8CpwEXPhoakqTpWDIAqup64P4DurcAu1t7N3DOov5LamAvcGySE4GXAFdX1f1V9QBwNT8eKpKkCVruNYATquqe1r4XOKG11wF3Lxq3r/Udqv/HJNmeZC7J3MLCwjLLkyQtZeSLwFVVQI2hlke3t7OqZqtqdmZmZlyblSQdYLkB8I12aof2+b7Wvx/YsGjc+tZ3qH5J0pQsNwD2AI/eybMVuGJR/2va3UCnAw+1U0VXAS9Ocly7+Pvi1idJmpI1Sw1IcinwAmBtkn0M7ua5CLgsyTbg68Ar2/ArgbOAeeBh4HUAVXV/kncBn2vj3llVB15YliRN0JIBUFXnHWLV5oOMLeD8Q2znYuDiw6pOkrRifCewJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSp5Z8GuhqtnHHp6ey3zsveulU9itJh8MjAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6tRIAZDkzUluTXJLkkuTPCnJSUluSDKf5JNJjm5jj2nL8239xnFMQJK0PMsOgCTrgN8DZqvqFOAo4FzgPcD7q+pZwAPAtvaSbcADrf/9bZwkaUpGPQW0BviJJGuAJwP3AC8ELm/rdwPntPaWtkxbvzlJRty/JGmZlh0AVbUf+GPgLga/+B8CbgQerKpH2rB9wLrWXgfc3V77SBt//HL3L0kazSingI5j8Ff9ScBPAU8Bzhi1oCTbk8wlmVtYWBh1c5KkQxjlFNCLgK9V1UJVfQ/4FPA84Nh2SghgPbC/tfcDGwDa+mcA3zpwo1W1s6pmq2p2ZmZmhPIkSY9llAC4Czg9yZPbufzNwG3AdcDL25itwBWtvact09ZfW1U1wv4lSSMY5RrADQwu5n4e+GLb1k7gbcAFSeYZnOPf1V6yCzi+9V8A7BihbknSiEb6j2BVdSFw4QHddwCnHWTst4FXjLI/SdL4+E5gSeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjo1UgAkOTbJ5Um+lOT2JM9N8swkVyf5avt8XBubJB9MMp/k5iSnjmcKkqTlGPUI4APAP1bVzwO/AtwO7ACuqapNwDVtGeBMYFP72A58eMR9S5JGsOwASPIM4PnALoCq+m5VPQhsAXa3YbuBc1p7C3BJDewFjk1y4rIrlySNZJQjgJOABeAvk9yU5KNJngKcUFX3tDH3Aie09jrg7kWv39f6JElTMEoArAFOBT5cVc8B/pf/P90DQFUVUIez0STbk8wlmVtYWBihPEnSYxklAPYB+6rqhrZ8OYNA+Majp3ba5/va+v3AhkWvX9/6fkRV7ayq2aqanZmZGaE8SdJjWXYAVNW9wN1Jfq51bQZuA/YAW1vfVuCK1t4DvKbdDXQ68NCiU0WSpAlbM+Lr3wh8PMnRwB3A6xiEymVJtgFfB17Zxl4JnAXMAw+3sZKkKRkpAKrqC8DsQVZtPsjYAs4fZX+SpPHxncCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROjRwASY5KclOSf2jLJyW5Icl8kk8mObr1H9OW59v6jaPuW5K0fOM4AngTcPui5fcA76+qZwEPANta/zbggdb//jZOkjQlIwVAkvXAS4GPtuUALwQub0N2A+e09pa2TFu/uY2XJE3BqEcAfwq8FfhBWz4eeLCqHmnL+4B1rb0OuBugrX+ojf8RSbYnmUsyt7CwMGJ5kqRDWXYAJHkZcF9V3TjGeqiqnVU1W1WzMzMz49y0JGmRNSO89nnA2UnOAp4EPB34AHBskjXtr/z1wP42fj+wAdiXZA3wDOBbI+xfkjSCZR8BVNXbq2p9VW0EzgWurapXAdcBL2/DtgJXtPaetkxbf21V1XL3L0kazUq8D+BtwAVJ5hmc49/V+ncBx7f+C4AdK7BvSdKQRjkF9ENV9Vngs619B3DaQcZ8G3jFOPYnSRqd7wSWpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROjeWdwJK0kjbu+PS0SzgieQQgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUqWUHQJINSa5LcluSW5O8qfU/M8nVSb7aPh/X+pPkg0nmk9yc5NRxTUKSdPhGOQJ4BHhLVZ0MnA6cn+RkYAdwTVVtAq5pywBnApvax3bgwyPsW5I0omUHQFXdU1Wfb+3/Bm4H1gFbgN1t2G7gnNbeAlxSA3uBY5OcuOzKJUkjGcs1gCQbgecANwAnVNU9bdW9wAmtvQ64e9HL9rW+A7e1PclckrmFhYVxlCdJOoiRAyDJU4G/BX6/qv5r8bqqKqAOZ3tVtbOqZqtqdmZmZtTyJEmHMFIAJHkig1/+H6+qT7Xubzx6aqd9vq/17wc2LHr5+tYnSZqCUe4CCrALuL2q3rdo1R5ga2tvBa5Y1P+adjfQ6cBDi04VSZImbJR/Cv884DeBLyb5Quv7Q+Ai4LIk24CvA69s664EzgLmgYeB142w78e1af4D6zsveunU9q3J8OdL47LsAKiqfwZyiNWbDzK+gPOXuz9J0niNcgQgqTPTPPrQ+PkoCEnqlAEgSZ0yACSpUwaAJHXKAJCkTnkX0BFmWndpeH+4tPp4BCBJnTIAJKlTBoAkdcprABqLHq89+K5YrXYGgFY1fwlLy+cpIEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMTD4AkZyT5cpL5JDsmvX9J0sBEAyDJUcCHgDOBk4Hzkpw8yRokSQOTPgI4DZivqjuq6rvAJ4AtE65BksTkA2AdcPei5X2tT5I0YY+7/weQZDuwvS3+T5Ivj7C5tcA3R69q1XHe/el17kfsvPOeJYc81tx/Zph9TDoA9gMbFi2vb30/VFU7gZ3j2FmSuaqaHce2VhPn3Z9e597rvGE8c5/0KaDPAZuSnJTkaOBcYM+Ea5AkMeEjgKp6JMkbgKuAo4CLq+rWSdYgSRqY+DWAqroSuHJCuxvLqaRVyHn3p9e59zpvGMPcU1XjKESStMr4KAhJ6tSqD4ClHi2R5Jgkn2zrb0iycfJVrowh5n5BktuS3JzkmiRD3Rr2eDfs40SS/EaSSnJE3CUyzLyTvLJ9z29N8teTrnGlDPGz/tNJrktyU/t5P2sadY5bkouT3JfklkOsT5IPtq/LzUlOPawdVNWq/WBwIfk/gJ8Fjgb+HTj5gDG/C3yktc8FPjntuic4918Hntzarz8S5j7MvNu4pwHXA3uB2WnXPaHv9ybgJuC4tvyT0657gnPfCby+tU8G7px23WOa+/OBU4FbDrH+LOAzQIDTgRsOZ/ur/QhgmEdLbAF2t/blwOYkmWCNK2XJuVfVdVX1cFvcy+B9F6vdsI8TeRfwHuDbkyxuBQ0z798GPlRVDwBU1X0TrnGlDDP3Ap7e2s8A/nOC9a2YqroeuP8xhmwBLqmBvcCxSU4cdvurPQCGebTED8dU1SPAQ8DxE6luZR3uYzW2MfhLYbVbct7tMHhDVX16koWtsGG+388Gnp3kX5LsTXLGxKpbWcPM/Y+AVyfZx+AuwzdOprSpG+nxOo+7R0Fo/JK8GpgFfm3atay0JE8A3ge8dsqlTMMaBqeBXsDgaO/6JL9UVQ9OtarJOA/4WFX9SZLnAn+V5JSq+sG0C3s8W+1HAEs+WmLxmCRrGBwefmsi1a2sYeZOkhcB7wDOrqrvTKi2lbTUvJ8GnAJ8NsmdDM6L7jkCLgQP8/3eB+ypqu9V1deArzAIhNVumLlvAy4DqKp/BZ7E4Fk5R7qhfg8cymoPgGEeLbEH2NraLweurXb1ZJVbcu5JngP8OYNf/kfK+eDHnHdVPVRVa6tqY1VtZHDt4+yqmptOuWMzzM/63zP4658kaxmcErpjkkWukGHmfhewGSDJLzAIgIWJVjkde4DXtLuBTgceqqp7hn3xqj4FVId4tESSdwJzVbUH2MXgcHCewcWUc6dX8fgMOff3Ak8F/qZd976rqs6eWtFjMOS8jzhDzvsq4MVJbgO+D/xBVa36o90h5/4W4C+SvJnBBeHXHgl/6CW5lEGor23XNy4EnghQVR9hcL3jLGAeeBh43WFt/wj4GkmSlmG1nwKSJC2TASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqf+DyyjeoB8cHzEAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist((coverage_array>0.5).mean(axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1129"
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum((coverage_array>0.5).mean(axis=1) > 0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "def sample_random_pair_of_genes(files):\n",
    "    while True:\n",
    "        file1, file2 = random.sample(files, 2)\n",
    "        id1 = int(file1.split('.')[0].split('-')[-1])\n",
    "        id2 = int(file2.split('.')[0].split('-')[-1])\n",
    "        if np.abs(id1-id2) < 15:\n",
    "            continue\n",
    "        if (id1 in core_genes) and (id2 in core_genes):\n",
    "            return file1, file2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('GUT_GENOME143712_1-1801.tsv', 'GUT_GENOME143712_1-1520.tsv')"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UHGG_utils.sample_random_pair_of_genes(gene_files, core_genes, gene_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GUT_GENOME000061'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header.split('\\t')[4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "microbiome",
   "language": "python",
   "name": "microbiome"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
